{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sqlite3\n",
    "import pandas as pd\n",
    "import os\n",
    "import sys\n",
    "from Bio.SubsMat import MatrixInfo as matlist\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio.Data import CodonTable\n",
    "from Bio.Seq import translate\n",
    "import math\n",
    "import numpy as np\n",
    "def fasta2List(pathFasta):\n",
    "    f = open(pathFasta, \"r\")\n",
    "    title = []\n",
    "    seq = []\n",
    "    seq_temp = []\n",
    "    for line in f:\n",
    "        if line[0] == \">\":\n",
    "            seq.append(''.join(seq_temp).replace(\"\\n\", \"\"))\n",
    "            title.append(line.replace(\"\\n\", \"\"))\n",
    "            seq_temp = []\n",
    "        else:\n",
    "            seq_temp.append(line)\n",
    "    seq.append(''.join(seq_temp).replace(\"\\n\", \"\"))\n",
    "    seq.pop(0)\n",
    "    dictionary = dict(zip(title, seq))\n",
    "    return dictionary\n",
    "\n",
    "conn = sqlite3.connect('../../mismatch_db.db')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "ename": "DatabaseError",
     "evalue": "Execution failed on sql 'SELECT transcript.*, protein.prot_ID, protein.sequence \nFROM transcript\nJOIN protein ON transcript.transcript_ID = protein.transcript_ID': no such table: transcript",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mOperationalError\u001b[0m                          Traceback (most recent call last)",
      "\u001b[0;32m~/anaconda3/envs/stage-env/lib/python3.7/site-packages/pandas/io/sql.py\u001b[0m in \u001b[0;36mexecute\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1585\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1586\u001b[0;31m             \u001b[0mcur\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexecute\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1587\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mcur\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mOperationalError\u001b[0m: no such table: transcript",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mDatabaseError\u001b[0m                             Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-48-194b4b18f0f3>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      6\u001b[0m \"\"\"SELECT transcript.*, protein.prot_ID, protein.sequence \n\u001b[1;32m      7\u001b[0m \u001b[0mFROM\u001b[0m \u001b[0mtranscript\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 8\u001b[0;31m JOIN protein ON transcript.transcript_ID = protein.transcript_ID\"\"\", conn)\n\u001b[0m\u001b[1;32m      9\u001b[0m \u001b[0mtranscript_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtranscript_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreindex\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"prot_ID\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"sequence\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"transcript_ID\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"sequence_genomic\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"sequence_CDS\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     10\u001b[0m \u001b[0;31m# transcript_df.to_csv(\"../../temp/nico/prot_transcript_seq.csv\", sep=\"\\t\", index=False)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/stage-env/lib/python3.7/site-packages/pandas/io/sql.py\u001b[0m in \u001b[0;36mread_sql_query\u001b[0;34m(sql, con, index_col, coerce_float, params, parse_dates, chunksize)\u001b[0m\n\u001b[1;32m    330\u001b[0m         \u001b[0mcoerce_float\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcoerce_float\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    331\u001b[0m         \u001b[0mparse_dates\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mparse_dates\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 332\u001b[0;31m         \u001b[0mchunksize\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mchunksize\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    333\u001b[0m     )\n\u001b[1;32m    334\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/stage-env/lib/python3.7/site-packages/pandas/io/sql.py\u001b[0m in \u001b[0;36mread_query\u001b[0;34m(self, sql, index_col, coerce_float, params, parse_dates, chunksize)\u001b[0m\n\u001b[1;32m   1631\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1632\u001b[0m         \u001b[0margs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_convert_params\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msql\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mparams\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1633\u001b[0;31m         \u001b[0mcursor\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexecute\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1634\u001b[0m         \u001b[0mcolumns\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mcol_desc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mcol_desc\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mcursor\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdescription\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1635\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/stage-env/lib/python3.7/site-packages/pandas/io/sql.py\u001b[0m in \u001b[0;36mexecute\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1596\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1597\u001b[0m             \u001b[0mex\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mDatabaseError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"Execution failed on sql '{args[0]}': {exc}\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1598\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mex\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mexc\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1599\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1600\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mstaticmethod\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mDatabaseError\u001b[0m: Execution failed on sql 'SELECT transcript.*, protein.prot_ID, protein.sequence \nFROM transcript\nJOIN protein ON transcript.transcript_ID = protein.transcript_ID': no such table: transcript"
     ]
    }
   ],
   "source": [
    "# Fichier pour Nicolas\n",
    "import sqlite3\n",
    "import pandas as pd\n",
    "conn = sqlite3.connect('mismatch_db.db')\n",
    "transcript_df = pd.read_sql_query(\n",
    "\"\"\"SELECT transcript.*, protein.prot_ID, protein.sequence \n",
    "FROM transcript\n",
    "JOIN protein ON transcript.transcript_ID = protein.transcript_ID\"\"\", conn)\n",
    "transcript_df = transcript_df.reindex([\"prot_ID\", \"sequence\", \"transcript_ID\", \"sequence_genomic\", \"sequence_CDS\"], axis=1)\n",
    "# transcript_df.to_csv(\"../../temp/nico/prot_transcript_seq.csv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "mismatch = pd.read_sql_query(\"SELECT * FROM mismatch\", conn)\n",
    "mismatch = mismatch.astype({\"exon_start_prim\": \"Int64\", \"exon_stop_prim\":\"Int64\", \"exon_start_hum\": \"Int64\", \"exon_stop_hum\":\"Int64\"})\n",
    "\n",
    "prim_exon_introns = pd.read_sql_query(\"\"\"\n",
    "SELECT mismatch_ID, mismatch.prot_prim, exon_intron_map.'type', exon_intron_map.number_elem, exon_intron_map.seq\n",
    "FROM mismatch\n",
    "JOIN protein ON mismatch.prot_prim = protein.prot_ID\n",
    "JOIN transcript ON protein.transcript_ID = transcript.transcript_ID\n",
    "JOIN exon_intron_map ON transcript.transcript_ID = exon_intron_map.transcript_ID\n",
    "\"\"\", conn)\n",
    "\n",
    "mismatch_prot_transcript = pd.read_sql_query(\n",
    "\"\"\"SELECT * \n",
    "FROM mismatch\n",
    "JOIN protein ON mismatch.prot_prim = protein.prot_ID\n",
    "JOIN transcript ON protein.transcript_ID = transcript.transcript_ID\"\"\", conn)\n",
    "\n",
    "mismatch_to_correct = pd.read_sql_query(\n",
    "\"\"\"SELECT mismatch.*\n",
    "FROM mismatch_flag \n",
    "JOIN tblastn_match ON mismatch_flag.mismatch_ID = tblastn_match.mismatch_ID \n",
    "JOIN mismatch on mismatch_flag.mismatch_ID = mismatch.mismatch_ID \n",
    "WHERE (one_hum_multiple_prim = 1\n",
    "OR non_canonical_hum_spl = 1\n",
    "OR N_in_genomic = 1\n",
    "OR small_introns = 1\n",
    "OR frameshift = 1)\n",
    "AND conserved = 0\n",
    "AND repeats_prot = 0\n",
    "AND alignement_error = 0\n",
    "AND human_isoform_exist = 0\n",
    "\"\"\", conn)\n",
    "\n",
    "tblastn = pd.read_sql_query(\n",
    "\"\"\"SELECT tblastn_match.*\n",
    "FROM tblastn_match \n",
    "JOIN mismatch_flag ON mismatch_flag.mismatch_ID = tblastn_match.mismatch_ID \n",
    "WHERE (one_hum_multiple_prim = 1\n",
    "OR non_canonical_hum_spl = 1\n",
    "OR N_in_genomic = 1\n",
    "OR small_introns = 1\n",
    "OR frameshift = 1)\n",
    "AND conserved = 0\n",
    "AND repeats_prot = 0\n",
    "AND alignement_error = 0\n",
    "AND human_isoform_exist = 0\n",
    "\"\"\", conn)\n",
    "\n",
    "protein_df = pd.read_sql_query(\n",
    "\"\"\"SELECT * \n",
    "FROM protein\n",
    "\"\"\", conn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/meyer/anaconda3/envs/stage-env/lib/python3.7/site-packages/ipykernel_launcher.py:38: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/home/meyer/anaconda3/envs/stage-env/lib/python3.7/site-packages/ipykernel_launcher.py:39: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    }
   ],
   "source": [
    "# SCRIPT DE CORRECTION DES MISMATCH\n",
    "total = 0\n",
    "error_big = 0\n",
    "error_fus = 0\n",
    "error_fus3 = 0\n",
    "total_OK = 0\n",
    "codon_stop = 0\n",
    "bug_bizarre = 0\n",
    "\n",
    "for index, row in mismatch_to_correct.iloc[:,:].iterrows():\n",
    "    transcript_map = prim_exon_introns.loc[(prim_exon_introns[\"mismatch_ID\"] == int(row[0])) & (prim_exon_introns[\"type\"] == \"Exon\")]\n",
    "    id_exon_start = row[7]-1\n",
    "    id_exon_stop = row[8]\n",
    "    mismatch_peptid = row[11]\n",
    "    \n",
    "    # LOCALISATION EXACT DES MISMATCH\n",
    "    mismatch_exon_table = transcript_map.iloc[id_exon_start:id_exon_stop]\n",
    "    mismatch_exon = transcript_map.iloc[id_exon_start:id_exon_stop, 4]\n",
    "    exon_seq = mismatch_exon.str.cat(sep=\"\")\n",
    "    for i in range(3):\n",
    "        peptid = translate(exon_seq[i:])\n",
    "        frame = i\n",
    "        match_find = peptid.find(row[11])\n",
    "        if match_find != -1:\n",
    "            break\n",
    "    start_nuc_mismatch = match_find*3+frame\n",
    "    stop_nuc_mismatch = match_find*3+len(mismatch_peptid*3)+frame\n",
    "    \n",
    "    # CREATION DES COORDONEE NUCLEOTIDES DES EXONS A MISMATCH\n",
    "    taille_exon = []\n",
    "    for index2, row2 in mismatch_exon_table.iloc[:,:].iterrows():\n",
    "        taille_exon.append(len(row2[4]))\n",
    "    taille_exon_fin = [sum(taille_exon[:x+1]) for x in range(0,len(taille_exon))]\n",
    "    taille_exon_debut = taille_exon_fin.copy()\n",
    "    taille_exon_debut.pop()\n",
    "    taille_exon_debut[:]=[i+1 for i in taille_exon_debut]\n",
    "    taille_exon_debut.insert(0, 0)\n",
    "    mismatch_exon_table[\"pos_start\"] = taille_exon_debut\n",
    "    mismatch_exon_table[\"pos_stop\"] = taille_exon_fin\n",
    "    \n",
    "    # SUPPRESSION DES PARTIE D'EXON DES TRANSCRIPT\n",
    "    exon_seq = transcript_map.iloc[:, 4].str.cat(sep=\"\")\n",
    "    #print(transcript_map)\n",
    "    #print(translate(exon_seq[:]))\n",
    "    #print(\"*\")\n",
    "    #print(row[0])\n",
    "    #print(frame)\n",
    "    #print(\"*\")\n",
    "    for index2, row2 in mismatch_exon_table.iloc[:,:].iterrows():\n",
    "        if row2[5] >= start_nuc_mismatch and row2[6] <= stop_nuc_mismatch:\n",
    "            #print(\"TO DELETE TOTALLY\", row2[3], index)\n",
    "            conditional_index = transcript_map.loc[ transcript_map['number_elem'] == row2[3] ].index\n",
    "            transcript_map.loc[conditional_index, [\"seq\"]] = \"\"\n",
    "\n",
    "        if row2[5] < start_nuc_mismatch and row2[6] < stop_nuc_mismatch:\n",
    "            #print(\"DELETE THE END OF EXON\", row2[3])\n",
    "            conditional_index = transcript_map.loc[ transcript_map['number_elem'] == row2[3] ].index\n",
    "            suppr = transcript_map.loc[conditional_index, \"seq\"].iloc[0][start_nuc_mismatch-row2[5]:]\n",
    "            transcript_map.loc[conditional_index, [\"seq\"]] = transcript_map.loc[conditional_index, \"seq\"].iloc[0][:start_nuc_mismatch-row2[5]]\n",
    "        if row2[5] > start_nuc_mismatch and row2[6] > stop_nuc_mismatch:\n",
    "            #print(\"DELETE THE BEGIN OF EXON\", row2[3])\n",
    "            conditional_index = transcript_map.loc[ transcript_map['number_elem'] == row2[3] ].index\n",
    "            suppr += transcript_map.loc[conditional_index, \"seq\"].iloc[0][:stop_nuc_mismatch-row2[5]+1]\n",
    "            transcript_map.loc[conditional_index, [\"seq\"]] = transcript_map.loc[conditional_index, \"seq\"].iloc[0][stop_nuc_mismatch-row2[5]+1:]\n",
    "\n",
    "        if row2[5] <= start_nuc_mismatch and row2[6] >= stop_nuc_mismatch:\n",
    "            # 1er nvx exon \n",
    "            conditional_index = transcript_map.loc[ transcript_map['number_elem'] == row2[3] ].index\n",
    "            transcript_map.loc[conditional_index, [\"seq\"]] = transcript_map.loc[conditional_index, \"seq\"].iloc[0][:start_nuc_mismatch-row2[5]]\n",
    "            # 2e nvx exon exon X.5\n",
    "            new_row = row2.copy()\n",
    "            new_row[3] = new_row[3] + 0.5\n",
    "            new_row[4] = new_row[4][stop_nuc_mismatch-new_row[5]:]\n",
    "            row_df = pd.DataFrame(new_row)\n",
    "            if len(row_df.index) > 1: row_df = row_df.T\n",
    "            transcript_map = pd.concat([transcript_map, row_df], ignore_index=True)\n",
    "            conditional_index = transcript_map.loc[ transcript_map['number_elem'] == row2[3]+0.5 ].index\n",
    "            \n",
    "    # CORRECTION\n",
    "    myTblastnMatch = tblastn.loc[ tblastn[\"mismatch_ID\"] == row[0] ]\n",
    "    \n",
    "    try:\n",
    "        prim_exon = prim_exon_introns.loc[(prim_exon_introns[\"mismatch_ID\"]==row[0]) & (prim_exon_introns[\"number_elem\"].isin(range(row[7],row[8]+1))) & (prim_exon_introns[\"type\"]==\"Exon\")]\n",
    "        prim_intron = prim_exon_introns.loc[(prim_exon_introns[\"mismatch_ID\"]==row[0]) & (prim_exon_introns[\"number_elem\"].isin(range(row[7],row[8]))) & (prim_exon_introns[\"type\"]==\"Intron\")]\n",
    "    except:\n",
    "        print(\"Error transcript\")\n",
    "    genomic_Seq = \"\"\n",
    "    for i in range(0,row[8]+1-row[7]):\n",
    "        genomic_Seq += prim_exon.iloc[i,4]\n",
    "        try: genomic_Seq += prim_intron.iloc[i,4]\n",
    "        except: pass\n",
    "    tblastn_dna = genomic_Seq[myTblastnMatch.iloc[0, 5]+2:myTblastnMatch.iloc[0, 6]]\n",
    "    \n",
    "    # Splicing site check\n",
    "    \"\"\"\n",
    "    no_splicing_site = False\n",
    "    splicing_5_pos = -1\n",
    "    finish_splicing = False\n",
    "    loop_value = [0, 1, 2, -1, -2]\n",
    "    for i in loop_value:\n",
    "        splicing_5 = genomic_Seq[myTblastnMatch.iloc[0, 5]-i*3:myTblastnMatch.iloc[0, 5]+2-i*3]\n",
    "        if  splicing_5 == \"AG\":\n",
    "            new_start = myTblastnMatch.iloc[0, 5]+2-i*3\n",
    "            for j in loop_value:\n",
    "                splicing_3 = genomic_Seq[myTblastnMatch.iloc[0, 6]+i*3:myTblastnMatch.iloc[0, 6]+2+i*3]\n",
    "                if  splicing_3 == \"GT\":\n",
    "                    new_stop = myTblastnMatch.iloc[0, 6]+i*3\n",
    "                    tblastn_dna = genomic_Seq[new_start:new_stop]\n",
    "                    finish_splicing = True\n",
    "                    break\n",
    "        if finish_splicing: break\n",
    "        else:\n",
    "            no_splicing_site = True\n",
    "    \n",
    "    if no_splicing_site: continue\n",
    "    \"\"\"\n",
    "    new_row_blast = row2.copy()\n",
    "    new_row_blast[3] = row[7] + 0.25\n",
    "    new_row_blast[4] = tblastn_dna\n",
    "    if \"*\" in translate(tblastn_dna): \n",
    "        continue\n",
    "    row_df_blast = pd.DataFrame(new_row_blast[:-2])\n",
    "    if len(row_df_blast.index) > 1: row_df_blast = row_df_blast.T\n",
    "    transcript_map = pd.concat([transcript_map, row_df_blast], ignore_index=True)\n",
    "\n",
    "    transcript_map = transcript_map[transcript_map['seq'].map(len) > 0]\n",
    "    transcript_map.sort_values(by=\"number_elem\", inplace=True, ignore_index=True)\n",
    "    transcript_map[\"number_elem\"] = range(1, len(transcript_map.index)+1)\n",
    "    exon_seq = transcript_map.iloc[:, 4].str.cat(sep=\"\")\n",
    "    \n",
    "    # VERIFIER QUE LE TRANSCRIPT S'ENCHAINE BIEN\n",
    "    exon_start = []\n",
    "    exon_stop = []\n",
    "    genomic = mismatch_prot_transcript.loc[mismatch_prot_transcript[\"mismatch_ID\"]== row[0] ].iloc[0, 19]\n",
    "    for index3, row3 in transcript_map.iloc[:,:].iterrows():\n",
    "        exon_start.append(genomic.find(row3[4]))\n",
    "        exon_stop.append(genomic.find(row3[4])+len(row3[4]))\n",
    "    transcript_map[\"exon_start\"] = exon_start\n",
    "    transcript_map[\"exon_stop\"] = exon_stop\n",
    "    \n",
    "    \n",
    "    total +=1\n",
    "    Error_switch = False\n",
    "    for i in range(len(transcript_map.index)-1):\n",
    "        if transcript_map.iloc[i, 6] > transcript_map.iloc[i+1, 5]:\n",
    "            if transcript_map.iloc[i, 6]-transcript_map.iloc[i+1, 5] > 15 and len(transcript_map.iloc[i, 4])>=17 and len(transcript_map.iloc[i+1, 4])>=17:\n",
    "                #print(\"Big error\")\n",
    "                #print(transcript_map.iloc[i, 4], len(transcript_map.iloc[i, 4]))\n",
    "                #print(transcript_map.iloc[i+1, 4], len(transcript_map.iloc[i+1, 4]))\n",
    "                error_big += 1\n",
    "                Error_switch = True\n",
    "                break\n",
    "            elif transcript_map.iloc[i, 6]-transcript_map.iloc[i+1, 5] <= 15 and len(transcript_map.iloc[i, 4])>=17 and len(transcript_map.iloc[i+1, 4])>=17:\n",
    "                #print(\"Fusion issue\")\n",
    "                #print(transcript_map.iloc[i, 6]-transcript_map.iloc[i+1, 5])\n",
    "                if (transcript_map.iloc[i, 6]-transcript_map.iloc[i+1, 5])%3 == 0: error_fus3 +=1\n",
    "                error_fus += 1\n",
    "                Error_switch = True\n",
    "                break\n",
    "    if Error_switch == True: continue\n",
    "    total_OK += 1\n",
    "    \n",
    "    # TRADUCTION DE PROT\n",
    "    # Mapping de l'ancienne CDS\n",
    "    old_cds = mismatch_prot_transcript.loc[mismatch_prot_transcript[\"mismatch_ID\"]== row[0] ].iloc[0, 18]\n",
    "    start_tri = exon_seq.find(old_cds[:45])\n",
    "    \n",
    "    # Si mapping foire on cherche la plus grand CDS contenant notre correction\n",
    "    if start_tri == -1:\n",
    "        frame = \"Error\"\n",
    "        start = []\n",
    "        stop = []\n",
    "        CDS = []\n",
    "        for i in range(3):\n",
    "            prot_seq = translate(exon_seq[i:])\n",
    "            if translate(tblastn_dna) in prot_seq:\n",
    "                frame = i\n",
    "                break\n",
    "        correction = prot_seq.find(translate(tblastn_dna))\n",
    "        if correction == -1: print(\"Error could not find tlbastn match in prot\")\n",
    "        old_prot = mismatch_prot_transcript.loc[mismatch_prot_transcript[\"mismatch_ID\"]== row[0] ].iloc[0, 14]\n",
    "        furthest_start = -1\n",
    "        stop_pos = -1\n",
    "        first_AA = -1\n",
    "        for i in range(correction,0,-1):\n",
    "            if prot_seq[i] == \"M\":\n",
    "                furthest_start = i\n",
    "            if prot_seq[i] == \"*\":\n",
    "                stop_pos = i\n",
    "                break\n",
    "        if furthest_start == -1 and stop_pos == -1: first_AA = 0\n",
    "        elif stop_pos < furthest_start: first_AA = furthest_start\n",
    "        elif stop_pos > furthest_start: first_AA = stop_pos+1\n",
    "\n",
    "        prot_seq = prot_seq[first_AA:]\n",
    "        stop_codon = prot_seq.find(\"*\")\n",
    "        if stop_codon != -1: prot_seq = prot_seq[:stop_codon]\n",
    "    \n",
    "    # Sinon si le mapping à fonctionné, on crée la prot.\n",
    "    else: \n",
    "        prot_seq = translate(exon_seq[start_tri:])\n",
    "        # Bug bizarre sur 3 prot pour 600 qui n'ont pas pu etre corrigé au final ? A voir\n",
    "        if translate(tblastn_dna) not in prot_seq:\n",
    "            bug_bizarre += 1\n",
    "            continue\n",
    "        \n",
    "        stop_codon = prot_seq.find(\"*\")\n",
    "        if stop_codon != -1: prot_seq = prot_seq[:stop_codon]\n",
    "\n",
    "  \n",
    "    #f = open(\"../../data/correction/two-by-two2/\"+str(row[0])+\"_\"+row[1]+\"_\"+row[2]+\".fasta\", \"w\")\n",
    "    f = open(\"../../data/correction/two-by-two2/\"+row[1]+\"_\"+row[2]+\".fasta\", \"w\")\n",
    "    f.write(\">\"+row[1]+\"\\n\")\n",
    "    f.write(protein_df.loc[protein_df[\"prot_ID\"] == row[1]].iloc[0, 1]+\"\\n\")\n",
    "    f.write(\">\"+row[2]+\"\\n\")\n",
    "    f.write(prot_seq+\"\\n\")\n",
    "    f.close()\n",
    "    \n",
    "    #f = open(\"../../data/correction/two-by-two-old2/\"+str(row[0])+\"_\"+row[1]+\"_\"+row[2]+\".fasta\", \"w\")\n",
    "    f = open(\"../../data/correction/two-by-two-old2/\"+row[1]+\"_\"+row[2]+\".fasta\", \"w\")\n",
    "    f.write(\">\"+row[1]+\"\\n\")\n",
    "    f.write(protein_df.loc[protein_df[\"prot_ID\"] == row[1]].iloc[0, 1]+\"\\n\")\n",
    "    f.write(\">\"+row[2]+\"\\n\")\n",
    "    f.write(protein_df.loc[protein_df[\"prot_ID\"] == row[2]].iloc[0, 1]+\"\\n\")\n",
    "    f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "from Bio import AlignIO\n",
    "import numpy as np\n",
    "\n",
    "identity_score = []\n",
    "old_identity_score = []\n",
    "coverage = []\n",
    "old_coverage = []\n",
    "\n",
    "alignements_files = glob.glob('../../data/correction/two-by-two2/*.mafft')\n",
    "alignements_files_old = glob.glob('../../data/correction/two-by-two-old2/*.mafft')\n",
    "for files_id in range(len(alignements_files)):\n",
    "    identity = 0\n",
    "    old_identity = 0\n",
    "    no_gap = 0\n",
    "    no_gap_old = 0\n",
    "    \n",
    "    align = AlignIO.read(alignements_files[files_id], \"fasta\")\n",
    "    for i in range(align.get_alignment_length()):\n",
    "        if align[0][i] != \"-\" and align[1][i] != \"-\":\n",
    "            no_gap += 1\n",
    "        if align[0][i] == align[1][i]:\n",
    "            identity += 1\n",
    "    #identity_score.append(identity/align.get_alignment_length()*100)\n",
    "    identity_score.append(identity/no_gap*100)\n",
    "    coverage.append(no_gap/align.get_alignment_length()*100)\n",
    "    \n",
    "    align_old = AlignIO.read(alignements_files_old[files_id], \"fasta\")\n",
    "    for i in range(align_old.get_alignment_length()):\n",
    "        if align_old[0][i] != \"-\" and align_old[1][i] != \"-\":\n",
    "            no_gap_old += 1\n",
    "        if align_old[0][i] == align_old[1][i]:\n",
    "            old_identity += 1\n",
    "    #old_identity_score.append(old_identity/align_old.get_alignment_length()*100)\n",
    "    old_identity_score.append(old_identity/no_gap_old*100)\n",
    "    old_coverage.append(no_gap_old/align_old.get_alignment_length()*100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "91.94731424216411\n",
      "95.03575670378338\n",
      "89.46483702547539\n",
      "90.72938944276979\n"
     ]
    }
   ],
   "source": [
    "print(np.mean(old_identity_score))\n",
    "print(np.mean(identity_score))\n",
    "print(np.mean(old_coverage))\n",
    "print(np.mean(coverage))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../../data/correction/two-by-two2/O14772_H2N716.fasta.mafft\n",
      "../../data/correction/two-by-two-old2/O14772_H2N716.fasta.mafft\n"
     ]
    }
   ],
   "source": [
    "diff = []\n",
    "for i in range(len(identity_score)):\n",
    "    if identity_score[i] - old_identity_score[i] > 30:\n",
    "        print(alignements_files[i])\n",
    "        print(alignements_files_old[i])\n",
    "    diff.append(identity_score[i] - old_identity_score[i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../../data/correction/two-by-two2/O14772_H2N716.fasta.mafft\n",
      "../../data/correction/two-by-two2/Q6ZNB5_H2NEJ2.fasta.mafft\n"
     ]
    }
   ],
   "source": [
    "diff2 = []\n",
    "for i in range(len(coverage)):\n",
    "    if coverage[i] - old_coverage[i] > 40:\n",
    "        print(alignements_files[i])\n",
    "    diff2.append(coverage[i] - old_coverage[i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAiwAAAGoCAYAAABlvr66AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8li6FKAAAgAElEQVR4nO3dfZwddX33/9cnu0FuxCKHmEKCRlm8R7CmiLVpEXZxoXLTolZrzemll1y9fpZE1KpY7WWv6/JXvbS1JL2Tn3i5ubQq2iKgsJBFVGhBCBUNELlcIZUQDOEAAkYhu/v9/TGz4eyyc/YkezOzu6/n47GPPd+ZOXM+55yZOe/znTkzkVJCkiSpyhaVXYAkSdJkDCySJKnyDCySJKnyDCySJKnyDCySJKnyDCySJKnyKhVYIiJFRNc+3ndVRNw53TW18bgviIjvRcSjEbFmgvFXRkS94L4r8ufcOUO1PRYRz5uJeVdJRCyNiO/k78FflV1PmSLi2fn73lF2LVUwH7cpqraI+K8RsSNfD2tl1zNbIuItEXH1jD7GvpyHJSK2AkuB4abBn0sp/cmUiolIwNEppcHpnHYmRcRFwCMppfP24b4rgLuBxSmloSnW8S3g8ymlz0xlPnNRRHwYeDlwdkopRcRHgK6U0h+WW9nMy9fF/5xSGii7lqlwmzKmjn3epqhcEbEYeAQ4IaX0/bLrmSnT+dm1N6byzf70ub6RnCbPAb5UdhFzSUR0TvNC/hzgjjRNZ0GcgfpG5xtkXxJGWg1bwNymZCq9TZmp9aMMM/BclgL7A7fvYz0dKaXhyafc6/k+5XnOyfcxpbTXf8BWoHuC4U8DHgZe2jRsCfAL4Fl5+x3AIPAgcBlwRNO0ieybMcC3yL45jo77I+D6/PZ38ml/DjwG/D5wIrCtafoX5fN4mGzhOaNp3OeAvwO+ATwKfBc4qsXzPSOfx8P5PF+UD/8m2TfCX+Z1PH+C++55HkAH8EngAeAu4J358+jMx/8KcBFwH3Av8D+Bjubnn9//IbJ0e2o+7qPj6vjb5tcT+HVgx+jj5OPOBm4teL6nAXfkr829wHubxp0J3Er2LeLHQG8+/Ij8/Xwwf3/f0XSfjwBfBT6f3+8/k+2O/EA+jwZwMXBoQT3PBL4O7Myf+9eB5U3v5W7gify5vy6/vTtvf7/N1/ZfgU/l9f/PCWroAD6Y1/socAtwZD7uN4CbgZ/l/39j3Pv/0Xz+v8jfj4mGFdbXtN5syR/7DuDXgP8DjOTzeAx4H7CCscvUZO/LxcCGfL63Ayv3ZZsw1T/cpuzNNuVQ4H8D28nWh6+NW06e8loA/wh8ctx8LgXe3bSc/DPZOnY3sGaS9fd44Ia8/vuAvwX2a7rPKcCdZOvE3wPfHvfav41seX4IuAp4TsHrtH/+uI38sW4Glu7r69C0TLwT+BFwdz7shcDGfPo7gTe2eO/+E0+ui3cB/yUf/vx8+Un5e/fNyeadLzf/AFyR33eidWA6n+dePXfgAOCvgP/I38vr82E/aXqejwGvoml9anO7+D/ItoGPAlcDh026nZjOjUs+7rPAR5va7wT689snkX1Y/xrZhmg98J293biMnzZvn0i+cQEW52/iB4H98sd9FHhB00LyINlK1wl8AfhSwfMZXQh78vm+L5/3fhPVOcH994wH/hj4IXBkvhBey9gPl68BnwYOAp4F3MSTK8MfkX0Iv4Psw/O/ki3AUVTHuNfzDvKAk7cvAd5TUPN9wKr89jOBX8tvH58vfD1kgWMZ8MJ83LfJNkz7A8eRbfhObtrg7QbOyu93APAu4EZgeb4sfBr4YkE9NbKAdSBwMPAVxq60n6MpZOSP9/lx85jstR0Czs2XhwMmqOFPgc3AC4AAjs3rOpRsI/LW/L5vztu1pvflJ8BL8vGLC4a1qu8NZCHm1/PH7iLfwDNuXeSpgWWy9+WXZAG1A/hL4MZ92SZM9W/883Cb0nKb8g3gy2Tr5mLgtyd7LYDfAu7hye3FM8lC3xFk6+QtwJ/nz+15ZB/Er22x/r4COCF/rivIPsDflU9/GFmw+b18/Nr8/qPbwbPy5/uifPyHgH8reK7/BbicbN3vyB/3Gfv6OjS9zxvJ1t0DyNa5e8iCSGd+vweAlxTU9DvAUWTr4m8Du3hyG7mCsetfy3mTLTc/A16dv7b7T8f7PdHz3JfnThbCv0W2re8gCyFPG/88x69PtLdd/DHZunBA3v7YpNuJKWxcHiNLvKN/78jHdQN3NU37r8Dq/PZFwP9qGvd0sgV5xTRvXFYBPwUWNY3/IvCRpoXkM03jTgN+WPBcPwxc3NReRPbhcWKbG5c948m+Pf1x07hTRt90sq7Ex2n6sMzf5Gubnv9g07gD8/v+alEd417P9wNfaFqYdgGHF9T8E7INxTPGDf808KkJpj+S7FvhwU3D/pLsGATINnjfGXefLeQfnHn78HxZ6JyopnH3PQ54qKn9OVoEljZf259M8ph3AmdOMPytwE3jht0A/FHT+/LfJ1gm/vte1HcVsLbFujhhYGnzfRloGvdi4BeTvf4z8YfblLa2KWTryQjwzAnGFb4WZB+uPwF+Kx/3Dp7sAXgl45Z/4HzgfzctJ9+ZqJ6m6d8FXJLfXg3c0DQuyD4UR7eDVwJvH/f8dzFBLwtZT8y/AS+bjteh6X0+qWn87wPXjZvHp4H/1uay+zXy9ZOnBpaW886Xmw0t5j1tz3Nvn3v+vvwCOHaCxx7zPMevT7S3XfxQ07j/h/xLSKu/qRzDclaaeH/zN4EDIuKVZCv4cWTf5iFL8/8+OmFK6bGIaJClt61TqGW8I4B70tjjAv4jf5xRP226vYvszS6a13+MNlJKIxFxz7h57VVd42oa9Ryy9HxfdlgDkC0wzdPvqTmltCufrqju8T4PbImIpwNvJFtI7yuY9myybz0fi4gfAB9IKd1A9gF4RcHzejCl9Oi457ayqX3P2LvwHOCSiGh+j4bJPrzvbZ4wIg4k213TS/YtA+Dgvdjf285rO76+8Y4k+0Yw3pjlIzd+WZto3s3DJquv6LEn0877Mn492L/EfdtuUyZ3JNl7+lDBfCd8LVJKWyPiS2RB+DvAH5BtEyBb/o6IiIeb5tUBXNfUHrMMR8Tzgb8mW5YOJAvItzTVsWf6lFKKiG1Nd38OcMG4X/QF2fMfvy79n/w5fykiDslr/rN9fR14cpkYv/69ctzz78wf+yki4lSyD/Tnk62nB5L1vk6knXm32vZM5/Oc6PFa1XcYWc/svm57Jtsutru+jClsWuUr38VkK8YO4OtNG8ztZC8QABFxEFm3+r1PmVHWZXpgU/tX96KM7cCREbGoaQPzbOD/7sU8mud1zGgjP0jySCaueTL35fcd9eym2/eQfcs+bB8/LFLLkSndGxE3AL9Lln7/ocW0NwNn5ke8/wnZcQ5H5jUeNcFdtgOHRsTBTe/1sxn7Go2v7x7gbSmlf21Vd+49ZLtiXplS+mlEHAd8j2wjN+FTmOCxJnttW75+PPncbxs3fMwynXs20D/JvJuHTVZf0eteNO/m2iZ7XyrPbcoY95C9p4eklB4eN26y1+KLwNUR8TGyXpXfbZrn3Smlo1s87vjl7B/I1sE3p5QejYh3Aa/Px91Htqu3+fktb7rvPWS7+L7Q8pkCKaXdwF8Af5H/MuUKst7OK9j312H887kH+HZKqWeyeiLiaWTH+qwGLk0p7Y6Ir1G8LWpn3q3W4am830Xzbuu5R8Qisl3GRwHjf/E02fayne3iXpup87D8E1lX01vy283D/1NEHJe/8f8v8N2U0tYJ5nEr8HsRcWB+HoW3jxu/g2xf60S+S7Zxel9ELI6IE4HT2bcj7y8GficiTs4/wN9D9uHyb/s4rzURsTwinkl20CkAeW/H1cBfRcQzImJRRBwVEb/d5rxbvR6jNpDtLz+GJ7+hjhER++W/p/+VfGPxCE/+1PQisvfv5Ly+ZRHxwpTSPWSvx19GxP4R8TKy96vVBukfgY9GxHPyx10SEWcWTHswWdfkwxFxKNm3m1Z2ACvyFW46XluAzwD/IyKOjszL8nMsXAE8PyL+ICI6I+L3yXatfL3dGbdR32eA90bEK/LH7hp93Wjxvu/j+1JVblPYs6xcCfx9RDwzr+W38tEtX4uU0vfIjmH6DHBV0wfgTcAjEfH+iDggIjoi4qUR8estSjmYbNvwWES8kOyYulHfAI6JiLMiO8fUOxkbDv8ROD8iXgIQEb8SEW+Y6EEi4jURcUxk5xV6hGyXx/BUXocJfJ1sHX5rPp/FEfHrEfGiCabdj+wYjp3AUGS9Lae0eJ32Zt5PMc3Pc6/qy4P5Z4G/jogj8uXiVflj7STbVVW0vkx5uziRqQSWyyM7Mc7o354PwJTS6Mp9BNmLPTr8GrL9t/9MlsKPAt5UMP9Pkf3aYwfQx1M3sh8B+iLi4Yh4Y/OIlNITZEfhn0p2ANHfk+3z/uHePsmU0p3AH5Id0PQA2Ubq9Pwx9tb/R3Y8wvfJuvL+Zdz41WQrxB1kByh9lWwfZjsuAF4fEQ9FxLqCaS4h3xWTUvp5i3m9FdgaEY+QHSj8hwAppZvIDs76FNmBYt/myRT9ZrL9mtvzx/lvKaWNk9R7Gdk3vkfJDsB9ZcG0f0N2YNYD+XSTpfSv5P8bETHaZTqV1xay7u+LyYLFI2Th7YCUUoPsl0nvIfslw/uA16WUHtiLebesL6X0FbJfFf0T2YGeXyM7DgmyY1I+lK8H751gvnv7vpTJbUp73kr2wf1D4H6y40fafS2+SHZM0J7Ql+9WPZ1sV9vdeU2fIfvlWpH3ku1WepRsu/blpvk9QHag+P8iWydeDGwiC2WklC4BPk62m+cRsl7LUwse51fJ1oVHyI57+zZP7sqayuuwR95bd0o+zXayXRUfJwsmE027hmxb8FD+Glw2HfNuYVqe5z7W916y3V03kx1U/nGy47h2kf/SMV9fThg33+naLo6xTyeO09wVET8m+/WJ57uQNOPyns5twFtSSteWXY/mrkqdml8zKyLOJtv3+M2ya5E0f0XEayPikHz3wQfJjvG4seSyNMfNyDVsVD2Rnbr/xcBbk2dVlTSzXkW222l0N+dZKaVflFuS5jp3CUmSpMpzl5AkSaq8Wd0ldNhhh6UVK1bM5kNKC9Ytt9zyQEppSdl1zDa3M9Lsmc3tzKwGlhUrVrBp06bZfEhpwYqI8WeaXBDczkizZza3M+4SkiRJlWdgkSRJlWdgkSRJldfWMSwRsZXsFMzDwFBKaWV+TZcvk532eyvwxoIrSkqSJE3J3vSwvCaldFxKafTS9B8Arsmv8HkNTRfykyRJmk5T2SV0JtkFxMj/nzX1ciRJkp6q3cCSyK6qe0tEnJMPW5pf+nr0EtjPmuiOEXFORGyKiE07d+6cesWSJGnBafc8LK9OKW2PiGcBGyOi7Uuqp5QuBC4EWLlypdcBkCRJe62tHpaU0vb8//3AJcDxwI6IOBwg/3//TBUpSZIWtkkDS0QcFBEHj94GTgFuAy4D6vlkdeDSmSpSkiQtbO3sEloKXBIRo9P/U0qpPyJuBi6OiLcDPwHeMHNlSpKkhWzSwJJSugs4doLhDeDkmShKkiSpmWe6VUuNRoM1a9bQaDTKLkXSPOV2Ru0wsKilvr4+Nm/ezIYNG8ouRdI85XZG7TCwqFCj0aC/v5+UEv39/X77kTTt3M6oXQYWFerr62NkZASA4eFhv/1ImnZuZ9QuA4sKDQwMMDQ0BMDQ0BAbN24suSJJ843bGbXLwKJC3d3ddHZmPyTr7Oykp6en5IokzTduZ9QuA4sK1et1Fi3KFpGOjg5Wr15dckWS5hu3M2qXgUWFarUavb29RAS9vb3UarWyS5I0z7idUbvavfihFqh6vc7WrVv91iNpxridUTsMLGqpVquxbt26ssuQNI+5nVE73CUkSZIqz8AiSZIqz8AiSZIqz8AiSSqVFz9UOwwskqRSefFDtcPAIkkqjRc/VLsMLJKk0njxQ7XLwCJJKo0XP1S7DCySpNJ48UO1y8AiSSqNFz9UuwwskqTSePFDtctrCUmSSuXFD9UOA4skqVRe/FDtcJeQJEmqPAOLJEmqPAOLJEmqPAOLJEmqPAOLJEmqPAOLJEmqPAOLJEmqPAOLJEmqPAOLJEmqPAOLpMqIiI6I+F5EfD1vPzcivhsRP4qIL0fEfmXXKKkcBhZJVbIW2NLU/jjwqZTS0cBDwNtLqUpS6QwskiohIpYDvwN8Jm8HcBLw1XySPuCscqqTVDYDi6Sq+BvgfcBI3q4BD6eUhvL2NmDZRHeMiHMiYlNEbNq5c+fMVypp1hlYJJUuIl4H3J9SuqV58ASTponun1K6MKW0MqW0csmSJTNSo6RydZZdgCQBrwbOiIjTgP2BZ5D1uBwSEZ15L8tyYHuJNUoqkT0skkqXUjo/pbQ8pbQCeBPwzZTSW4Brgdfnk9WBS0sqUTOo0WiwZs0aGo1G2aWowgwskqrs/cC7I2KQ7JiWi0quRzOgr6+PzZs3s2HDhrJLUYUZWCRVSkrpWyml1+W370opHZ9S6kopvSGl9HjZ9Wl6NRoN+vv7SSnR399vL4sKGVgkSaXp6+tjZCT7Ydjw8LC9LCpkYJEklWZgYIChoeyX60NDQ2zcuLHkilRVBhZJUmm6u7vp7Mx+sNrZ2UlPT0/JFamqDCySpNLU63UWLco+ijo6Oli9enXJFamqDCySpNLUajV6e3uJCHp7e6nVamWXpIryxHGSpFLV63W2bt1q74paMrBIkkpVq9VYt25d2WWo4twlJEmSKs/AIkmSKs/AIkmSKs/AIkmSKs/AIkmSKs/AIkmSKs/AIkmSKs/AIkmSKs/AIkmSKs/AIkmSKs/AIkmSKs/AIkkqVaPRYM2aNTQajbJLUYUZWCRJperr62Pz5s1s2LCh7FJUYQYWSVJpGo0G/f39pJTo7++3l0WFDCySpNL09fUxMjICwPDwsL0sKtR2YImIjoj4XkR8PW8/NyK+GxE/iogvR8R+M1emJGk+GhgYYGhoCIChoSE2btxYckWqqr3pYVkLbGlqfxz4VErpaOAh4O3TWZgkaf7r7u4mIgCICHp6ekquSFXVVmCJiOXA7wCfydsBnAR8NZ+kDzhrJgqUJM1fZ5xxBiklAFJKnH766SVXpKpqt4flb4D3ASN5uwY8nFIaytvbgGUT3TEizomITRGxaefOnVMqVpI0v1x22WVjelguv/zykitSVU0aWCLidcD9KaVbmgdPMGma6P4ppQtTSitTSiuXLFmyj2VKkuajgYGBMT0sHsOiIu30sLwaOCMitgJfItsV9DfAIRHRmU+zHNg+IxVKkuat7u5uOjuzj5LOzk6PYVGhSQNLSun8lNLylNIK4E3AN1NKbwGuBV6fT1YHLp2xKiVJ81K9XmfRouyjqKOjg9WrV5dckapqKudheT/w7ogYJDum5aLpKUmStFDUajV6e3uJCHp7e6nVamWXpIrqnHySJ6WUvgV8K799F3D89JckSVpI6vU6W7dutXdFLe1VYJEkabrVajXWrVtXdhmqOE/NL0mSKs/AIkmSKs/AIkmSKs/AIkmSKs/AIkmSKs/AIkmSKs/AIkmSKs/AIkmSKs/AIkmSKs/AIkmSKs/AIkmSKs/AIkmSKs/AIkmSKs/AIkmSKs/AIkmSKs/AIqkSImL/iLgpIr4fEbdHxF/kw58bEd+NiB9FxJcjYr+ya9X0ajQarFmzhkajUXYpqjADi6SqeBw4KaV0LHAc0BsRJwAfBz6VUjoaeAh4e4k1agb09fWxefNmNmzYUHYpqjADi6RKSJnH8ubi/C8BJwFfzYf3AWeVUJ5mSKPRoL+/n5QS/f399rKokIFFUmVEREdE3ArcD2wEfgw8nFIayifZBiyb4H7nRMSmiNi0c+fO2StYU9bX18fIyAgAw8PD9rKokIFFUmWklIZTSscBy4HjgRdNNNkE97swpbQypbRyyZIlM12mptHAwABDQ1keHRoaYuPGjSVXpKoysEiqnJTSw8C3gBOAQyKiMx+1HNheVl2aft3d3XR2Zm9vZ2cnPT09JVekqjKwSKqEiFgSEYfktw8AuoEtwLXA6/PJ6sCl5VSomVCv11m0KPso6ujoYPXq1SVXpKoysEiqisOBayPiB8DNwMaU0teB9wPvjohBoAZcVGKNmma1Wo3e3l4igt7eXmq1WtklqaI6J59EkmZeSukHwMsnGH4X2fEsmqfOOOMMrrnmGk4//fSyS1GF2cMiSSrVZZddxq5du7j88svLLkUVZmCRJJXG87CoXQYWSVJpPA+L2mVgkSSVxvOwqF0GFklSaTwPi9plYJEklcbzsKhdBhZJUmk8D4va5XlYJEmlqtfrbN261d4VtWRgkSSVqlarsW7durLLUMW5S0iSJFWegUWSJFWegUWSJFWegUWSVKpGo8GaNWs8Lb9aMrBIkkrV19fH5s2bPS2/WjKwSJJK48UP1S4DiySpNF78UO0ysEiSSuPFD9UuA4skqTRe/FDtMrBIkkrjxQ/VLgOLJKk0XvxQ7fJaQpKkUnnxQ7XDwCJJKpUXP1Q73CUkSZIqz8AiSZIqz8AiSZIqz8AiSZIqz8AiSZIqz8AiSZIqz8AiSZIqz8AiSZIqz8AiSZIqz8AiSZIqz8AiSZIqz8AiSSpVo9FgzZo1NBqNsktRhRlYJEml6uvrY/PmzWzYsKHsUlRhBhZJUmkajQZXXnklKSWuvPJKe1lUyMAiSSpNX18fQ0NDAOzevdteFhUysEiSSrNx40ZSSgCklLj66qtLrkhVNWlgiYj9I+KmiPh+RNweEX+RD39uRHw3In4UEV+OiP1mvlxJ0nyydOnSlm1pVDs9LI8DJ6WUjgWOA3oj4gTg48CnUkpHAw8Bb5+5MiVJ89GOHTtatqVRkwaWlHksby7O/xJwEvDVfHgfcNaMVChJmrd6enqICAAiglNOOaXkilRVbR3DEhEdEXErcD+wEfgx8HBKaSifZBuwrOC+50TEpojYtHPnzumoWZI0T9TrdRYvXgzA4sWLWb16dckVqaraCiwppeGU0nHAcuB44EUTTVZw3wtTSitTSiuXLFmy75VKkuadWq1Gb28vEcGpp55KrVYruyRVVOfeTJxSejgivgWcABwSEZ15L8tyYPsM1CdJmufq9Tpbt261d0UttfMroSURcUh++wCgG9gCXAu8Pp+sDlw6U0VKkuavWq3GunXr7F1RS+3sEjocuDYifgDcDGxMKX0deD/w7ogYBGrARTNXpiRpvvJaQmrHpLuEUko/AF4+wfC7yI5nkSRpnzVfS+i8884ruxxVlGe6lVQJEXFkRFwbEVvyk1SuzYcfGhEb85NUboyIZ5Zdq6ZPo9Ggv7+flBL9/f32sqiQgUVSVQwB70kpvYjswP53RsSLgQ8A1+Qnqbwmb2ue6OvrY2RkBIDh4WGvJaRCBha15L5lzZaU0n0ppX/Pbz9KdnD/MuBMspNTgiepnHcGBgb2XPxwaGiIjRs3llyRqsrAopaa9y1LsyUiVpAdO/ddYGlK6T7IQg3wrAmm9wSVc1R3dzedndnhlJ2dnfT09JRckarKwKJC7ltWGSLi6cA/A+9KKT3Szn08QeXcVa/XWbQo+yhatGiR52JRIQOLCrlvWbMtIhaThZUvpJT+JR+8IyIOz8cfTnaJEM0TtVqNI444AoAjjjjCc7GokIFFhdy3rNkU2RXwLgK2pJT+umnUZWQnpwRPUjnvNBoN7r33XgC2b99uT64KGVhUyH3LmmWvBt4KnBQRt+Z/pwEfA3oi4kdAT97WPNHX10dK2aXoRkZG7MlVIQOLCtXr9T2XfXffsmZaSun6lFKklF6WUjou/7sipdRIKZ2cUjo6//9g2bVq+tiTq3YZWFSoVquxbNkywH3LkmaGPblql4FFhRqNBtu3Zxfhdt+ypJnQ/Cuhjo4Oe3JVyMCiQs2/EnLfsqSZUKvV6O3tJSLo7e21J1eFDCwq5L5lSbOhXq9zzDHH2LuilgwsKrRq1aqWbUmaDrVajXXr1tm7opYMLCo0+lNDSZLKZmBRoeuvv35M+7rrriupEknzmRdZVTsMLCrU3d1NR0cHkB29788NJc0EL7KqdhhYVKher+8JLJ2dnR4QJ2naeZFVtcvAokL+3FDSTPMiq2qXgUUt+XNDSTPJ0yeoXQYWteTPDSXNJE+foHYZWNSSR+9Lmkm//OUvx7Qff/zxkipR1RlY1JJH70uaSZ4+Qe0ysKiQR+9LkqrCwKJCHr0vaaYdccQRLdvSKAOLCnn0vqSZ9sADD7RsS6MMLCrU3d1NZ2cnkJ04zjPdSppu47crp5xySkmVqOoMLCpUr9dZtChbRDo6OjwXi6Rpd8YZZ4xpn3766SVVoqozsKiQZ7qVNNMuu+wyIgKAiODyyy8vuSJVlYFFLXmmW0kzaWBggJQSACklj5VTIQOLWvJMt5JmksfKqV0GFklSaer1+p5dQosWLbI3V4UMLJKk0tRqNZYtWwZk52CxN1dFDCySpNI0Gg22b98OwPbt2z2jtgoZWCRJpWk+o/bIyIhn1FYhA4skqTSeUVvtMrCopUajwZo1a+ymlTQjVq1a1bItjTKwqKW+vj42b95sN62kGTF6DhZpMgYWFWo0GvT395NSor+/314WSdPu+uuvH9O+7rrrSqpEVWdgUaHmg+GGh4ftZZE07bq7u+no6ACya5Z54jgVMbCokAfDSZpp9Xp9T2Dp7Oz0xHEqZGBRIb/5SJppXmRV7TKwqFC9Xh9zUTK/+UiaCV5kVe0wsEiSSvXQQw/x4x//mIceeqjsUlRhBhYV6uvrY9GibBFZtGiRB91KmhEf/vCH+fnPf86f//mfl12KKszAokIedCtppg0ODnLfffcB2bWEBgcHS65IVWVgUaHu7m46OzuB7Oh9D7qVNN0+/OEPj2nby6IiBhYVqtfre3YJdXR0eECcpGk32rsyavTKzdJ4BhYV8ueGkqSq6Cy7AFVbvV5n69at9q5IkkplYFFLtVqNdevWlV2GJGmBc5eQJEmqPAOLJKk0EdGyLY0ysEiSSjN6+Y+itjTKwCJJKs3oBVaL2tIoA4ukSoiIz0bE/RFxW9OwQyNiY0T8KIZxJjgAABI1SURBVP//zDJr1PQ77rjjxrRf/vKXl1SJqs7AIqkqPgf0jhv2AeCalNLRwDV5W/PIli1bxrTvuOOOkipR1RlYJFVCSuk7wIPjBp8J9OW3+4CzZrUozbhdu3a1bEujDCySqmxpSuk+gPz/syaaKCLOiYhNEbFp586ds1qgpNlhYJE056WULkwprUwprVyyZEnZ5WgvnHDCCWPar3rVq0qqRFVnYJFUZTsi4nCA/P/9Jdejafanf/qnY9rvfe97S6pEVWdgkVRllwH1/HYduLTEWiSVyMAiqRIi4ovADcALImJbRLwd+BjQExE/AnrytuaRCy+8sGVbGjVpYImIIyPi2ojYEhG3R8TafLjnR1gAGo0Ga9asodFolF2K5rmU0ptTSoenlBanlJanlC5KKTVSSienlI7O/4//FZHmuIGBgZZtaVQ7PSxDwHtSSi8CTgDeGREvxvMjLAh9fX1s3ryZDRs2lF2KpHloeHi4ZVsaNWlgSSndl1L69/z2o8AWYBmeH2HeazQaXHnllaSUuPLKK+1lkTTtPDW/2rVXx7BExArg5cB38fwI815fXx9DQ0MA7N69214WSdNu1apVLdvSqLYDS0Q8Hfhn4F0ppUfavZ/nR5i7Nm7cuOfKqSklrr766pIrkiQtVG0FlohYTBZWvpBS+pd8sOdHmOeWLl3asi1JU3Xddde1bEuj2vmVUAAXAVtSSn/dNMrzI8xzO3bsaNmWpKnyoFu1q50ellcDbwVOiohb87/T8PwI815PTw9ZXoWI4JRTTim5IknSQtU52QQppeuBKBh98vSWoyqp1+v09/fzxBNPsHjxYlavXl12SZKkBcoz3apQrVajt7eXiODUU0+lVquVXZIkaYGatIdFC1u9Xmfr1q32rkiSSmVgUUu1Wo1169aVXYakeSoi9pw+YbQtTcRdQpKk0ixatKhlWxrlkiFJKs1LXvKSMe2XvvSlJVWiqjOwSJJKc/vtt49p33bbbSVVoqozsEiSSuOJ49QuA4skSao8A4skSao8A4taajQarFmzhkajUXYpkqQFzMCilvr6+ti8eTMbNmwouxRJ0gJmYFGhRqNBf38/KSX6+/vtZZEklcbAokJ9fX2MjIwA2ZH79rJIkspiYFGhgYEBhoaGABgaGmLjxo0lVyRJWqgMLCq0atWqlm1JkmaLFz9UoeYLkkmaO9avX8/g4GDZZeyztWvXll3CpLq6ujj33HPLLmNBsYdFha6//vox7euuu66kSiTNVwceeOCY9kEHHVRSJao6e1hUqLu7m2984xsMDw/T0dFBT09P2SVJasNc+ubfaDQ4++yz97Q3bNhArVYrsSJVlT0sKlSv1+no6ACgs7OT1atXl1yRpPmmVqvt6WV5xSteYVhRIQOLCtVqNXp7e4kIent73ZBImhErVqzgoIMO4oMf/GDZpajC3CWklur1Olu3brV3RdKMWbx4MV1dXX4pUksGFrVUq9VYt25d2WVIkhY4dwmpJS9+KEmqAgOLWvLih5KkKjCwqJAXP5QkVYWBRYX6+voYHh4GsmsJ2csiSSqLgUWFBgYG9gSW4eFhL34oSSqNgUWFjj/++JZtSZJmi4FFhcZfPG0uX0xNkjS3GVhUaNu2bS3bkiTNFgOLCq1YsaJlW5Kk2WJgUaEPfehDLduSJM0WA4sKdXV1cfjhhwNwxBFH0NXVVXJFkqSFymsJqaWRkRGAPT9vlhai9evXe9D5DBp9bdeuXVtyJfNXV1cX5557btllTImBRYUGBwfZsWMHADt27GBwcNBeFi1Ig4OD3HrbFoYPPLTsUualRU8kAG65a0fJlcxPHbseLLuEaWFgUaGPfOQjT2l//vOfL6cYqWTDBx7KL154WtllSHvtgB9eUXYJ08JjWFTInzVLkqrCwCJJkirPwKJCHR0dLduSJM0WA4sK/eZv/uaY9qpVq0qqRAtdRPRGxJ0RMRgRHyi7Hkmzz4NuVWj//fcf037a055WUiVayCKiA/g7oAfYBtwcEZellO6YrRruvfdeOnb9bN4cvKiFpWNXg3vvHSq7jCmzh0WFrrvuupZtaZYcDwymlO5KKT0BfAk4s+SaJM0ye1hUaNWqVVx11VVj2lIJlgH3NLW3Aa9sniAizgHOAXj2s589/QUsW8ZPH+/0Z82akw744RUsW7a07DKmzB4WFUoplV2CBBATDBuzcKaULkwprUwprVyyZMkslSVpNhlYVOj6668f03aXkEqyDTiyqb0c2F5SLZJK4i4hFXKXkCriZuDoiHgucC/wJuAPZruIjl0PetDtDFn0y0cAGNn/GSVXMj9lp+af+7uEDCwq5C4hVUFKaSgi/gS4CugAPptSun02a/AaWjNrcPBRALqeN/c/VKtp6bxYhg0sKjTRLqHzzz+/pGq0kKWUrgBK696Y61e5rbrRqzRfcMEFJVeiKvMYFhXyxHGSpKowsKjQE088Mab9+OOPl1SJJGmhM7Co0PhdQuPbkiTNFgOLCo0/6NaDcCVJZTGwqJDHsEiSqsLAokIRE51gVJKk2WdgUSGPYZEkVYWBRYWGh4dbtiVJmi0GFhXq6Oho2ZYkabYYWFRoaGioZVuSpNliYFGh8QfdehCuJKksBhYV8jwskqSqMLCo0JFHHtmyLUnTYffu3QwODtJoNMouRRVmYFGhZcuWjWkvX768pEokzWfbtm3j5z//OevXry+7FFWYgUWFbr755jHtm266qaRKJM1XjUaDRx55BIBvf/vb9rKoUOdkE0TEZ4HXAfenlF6aDzsU+DKwAtgKvDGl9NDMlakyeNCtNDetX7+ewcHBsstoy913373ndkqJt73tbaxYsaK8gtrU1dXFueeeW3YZC0o7PSyfA3rHDfsAcE1K6Wjgmryteebkk09u2ZakqRrtXRn1s5/9rKRKVHWT9rCklL4TESvGDT4TODG/3Qd8C3j/NNalCnjDG97AVVddNaYtqfrm0jf/E0888SnDLrjggtkvRJW3r8ewLE0p3QeQ/3/W9JWkqrj44ovHtL/yla+UVIkkaaGb8YNuI+KciNgUEZt27tw50w+naXTNNdeMaQ8MDJRUiSRpodvXwLIjIg4HyP/fXzRhSunClNLKlNLKJUuW7OPDqQwjIyMt25IkzZZ9DSyXAfX8dh24dHrKUZV4pltJUlVMGlgi4ovADcALImJbRLwd+BjQExE/AnrytiRJ0oxo51dCby4Y5W9cJUnSrPBMt5IkqfIMLJIkqfIMLCpUq9XGtA877LCSKpEkLXQGFhV67LHHxrQfffTRkiqRNF+97GUvG9M+9thjS6pEVWdgUaGOjo6WbUmaqjvuuGNM+/bbby+pElWdgUWFdu3a1bItSVM1NDTUsi2NMrBIkkrT2dnZsi2NMrCo0LOeNfaalkuXLi2pEknz1fhjVjyGRUUMLCp01FFHjWk/73nPK6kSSfPVli1bWralUQYWFbrxxhtbtiVpqrq7u4kIACKCnp6ekitSVRlYVMiLH0qaafV6ncWLFwOw3377sXr16pIrUlUZWCRJpanVarzmNa8B4MQTT3zKCSulUR6OPcvWr1/P4OBg2WXss7Vr15ZdwqS6uro499xzyy5DUpvsvVU77GFRoYMPPrhlW5KmqtFocO211wLwzW9+k0ajUXJFqip7WGbZXPrm32g0OPvss/e0P/e5z9ldK2la9fX1sXv3bgB2797Nhg0bOO+880quSlVkD4sK1Wq1Pb0qv/Ebv2FYkTTtrr766jHtq666qqRKVHUGFrV05JFHctBBB/Ge97yn7FIkzUOe6VbtMrCopcWLF9PV1WXviqQZMf6q8OPb0igDiySpNE9/+tNbtqVRBhZJUmm8WrPaZWCRJJXmlFNOGdN+7WtfW1IlqjoDiySpNM2n5l+8eLGn5lchA4skqTS1Wo1TTz2ViOC0007zAH8VMrBIKl1EvCEibo+IkYhYOW7c+RExGBF3RoT7C+aher3OMcccY++KWvIH75Kq4Dbg94BPNw+MiBcDbwJeAhwBDETE81NKw7NfomZKrVZj3bp1ZZehirOHRVLpUkpbUkp3TjDqTOBLKaXHU0p3A4PA8bNbnaQqMLBIqrJlwD1N7W35sDEi4pyI2BQRm3bu3DlrxUmaPQYWSbMiIgYi4rYJ/s5sdbcJhqWnDEjpwpTSypTSyiVLlkxf0ZoVjUaDNWvWeKVmtWRgkTQrUkrdKaWXTvB3aYu7bQOObGovB7bPbKWabZ/4xCf4wQ9+wCc/+cmyS1GFGVgkVdllwJsi4mkR8VzgaOCmkmvSNGo0Gtx4440A3HDDDfayqJCBRVLpIuJ3I2Ib8CrgGxFxFUBK6XbgYuAOoB94p78Qml8+8YlPjGnby6Ii/qxZUulSSpcAlxSM+yjw0dmtSLNltHdl1A033FBSJao6e1gkSVLlGVgkSaU58MADW7alUQYWSVJpjj322JZtaZSBRZJUmltvvbVlWxplYJEklWbp0qUt29IoA4skqTQ7duxo2ZZGGVgkSaXp6ekhIrsCQ0RwyimnlFyRqsrAIkkqTb1ep7MzOyXY4sWLWb16dckVqarm/Inj1q9fz+DgYNllzFujr+3atWtLrmT+6urq4txzzy27DKkUtVqNU089lcsvv5xTTz2VWq1WdkmqqDkfWAYHB7n1ti0MH3ho2aXMS4ueyC6Me8td7leeCR27Hiy7BKl09XqdrVu32ruiluZ8YAEYPvBQfvHC08ouQ9prB/zwirJLkEpXq9VYt25d2WWo4jyGRZIkVZ6BRZIkVZ6BRZIkVZ6BRZIkVZ6BRZIkVZ6BRZIkVZ6BRZIkVZ6BRZIkVZ6BRZIkVZ6BRZIkVd6cPzX/vffeS8eun3mKc81JHbsa3HvvUNllSFLl2cMiSZIqb873sCxbtoyfPt7pxQ81Jx3wwytYtmxp2WVIUuXZwyJJkirPwCJJkirPwCJJkipvzh/DAtCx60F/JTRDFv3yEQBG9n9GyZXMTx27HgQ8hkWSJjPnA0tXV1fZJcxrg4OPAtD1PD9UZ8ZSl2FJasOcDyznnntu2SXMa2vXrgXgggsuKLkSSdJC5jEskiSp8gwskiSp8qYUWCKiNyLujIjBiPjAdBUlSZLUbJ8DS0R0AH8HnAq8GHhzRLx4ugqTJEkaNZUeluOBwZTSXSmlJ4AvAWdOT1mSJElPmkpgWQbc09Telg8bIyLOiYhNEbFp586dU3g4SZK0UE0lsMQEw9JTBqR0YUppZUpp5ZIlS6bwcJKk+ajRaLBmzRoajUbZpajCphJYtgFHNrWXA9unVo4kaaHp6+tj8+bNbNiwoexSVGFTCSw3A0dHxHMjYj/gTcBl01OWJGkhaDQa9Pf3k1Kiv7/fXhYV2ucz3aaUhiLiT4CrgA7gsyml26etsnlq/fr1DA4Oll1G20ZrHT3j7VzQ1dXlGZClOaKvr4+RkREAhoeH2bBhA+edd17JVamKpnQelpTSFSml56eUjkopfXS6ilJ1HHDAARxwwAFllyFpnhoYGGBoaAiAoaEhNm7cWHJFqqo5fy2hucZv/pL0pO7ubq644gqGhobo7Oykp6en7JJUUZ6aX5JUmnq9zqJF2UdRR0cHq1evLrkiVZWBRZJUmlqtRm9vLxFBb28vtVqt7JJUUe4SkiSVql6vs3XrVntX1JKBRZJUqlqtxrp168ouQxXnLiFJpYuIT0TEDyPiBxFxSUQc0jTu/PyK8HdGxGvLrFNSeQwskqpgI/DSlNLLgP8LnA+QXwH+TcBLgF7g7/MrxUtaYAwskkqXUro6pTSUN28ku9QHZFeA/1JK6fGU0t3AINmV4iUtMAYWSVXzNuDK/LZXhZcEeNCtpFkSEQPAr04w6s9SSpfm0/wZMAR8YfRuE0w/4VXhgQsBVq5c+ZTxkuY+A4ukWZFS6m41PiLqwOuAk1NKo6HDq8JLAtwlJKkCIqIXeD9wRkppV9Ooy4A3RcTTIuK5wNHATWXUKKlc9rBIqoK/BZ4GbIwIgBtTSn+cUro9Ii4G7iDbVfTOlNJwiXVKKomBRVLpUkpdLcZ9FPBq8NIC5y4hSZJUeQYWSZJUefHkwfiz8GARO4H/mLUH1HQ5DHig7CK0156TUlpSdhGzze3MnOV2Zm6ate3MrAYWzU0RsSmltLLsOiTNX25nNBl3CUmSpMozsEiSpMozsKgdF5ZdgKR5z+2MWvIYFkmSVHn2sEiSpMozsEiSpMozsEiSpMozsEiSpMozsEiSpMr7/wFkcK2IOg8qpQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 576x432 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "f, (ax1, ax2) = plt.subplots(1, 2, figsize=(8, 6))\n",
    "sns.boxplot(diff, orient=\"v\", ax=ax1)\n",
    "sns.boxplot(diff2, orient=\"v\", ax=ax2)\n",
    "ax1.set_title(\"Evolution of identiy score after correction\")\n",
    "ax2.set_title(\"Evolution of coverage score afer correction\")\n",
    "f.tight_layout()\n",
    "plt.savefig(\"../../results/figures/boxplot2.png\", dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "mismatch_to_correct = pd.read_sql_query(\n",
    "\"\"\"SELECT mismatch.*\n",
    "FROM mismatch_flag\n",
    "JOIN mismatch on mismatch_flag.mismatch_ID = mismatch.mismatch_ID \n",
    "WHERE (one_hum_multiple_prim = 1\n",
    "OR non_canonical_hum_spl = 1\n",
    "OR N_in_genomic = 1\n",
    "OR small_introns = 1)\n",
    "AND conserved = 0\n",
    "AND repeats_prot = 0\n",
    "AND alignement_error = 0\n",
    "AND human_isoform_exist = 0\n",
    "\"\"\", conn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "mismatch_to_correct\n",
    "mismatch_to_correct.to_csv('../../temp/primate_errors_misprediction.csv', sep=\"\\t\", index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compte nosmascus\n",
    "conn = sqlite3.connect('../../mismatch_db.db')\n",
    "mismatch = pd.read_sql_query(\n",
    "\"\"\"SELECT *\n",
    "FROM mismatch\n",
    "JOIN protein ON mismatch.prot_prim = protein.prot_ID\n",
    "\"\"\", conn)\n",
    "mismatch = mismatch.astype({\"exon_start_prim\": \"Int64\", \"exon_stop_prim\":\"Int64\", \"exon_start_hum\": \"Int64\", \"exon_stop_hum\":\"Int64\"})\n",
    "\n",
    "\n",
    "\n",
    "mismatch_mispredicted = pd.read_sql_query(\n",
    "\"\"\"SELECT mismatch.*, protein.*\n",
    "FROM mismatch_flag \n",
    "JOIN protein ON mismatch.prot_prim = protein.prot_ID\n",
    "JOIN mismatch on mismatch_flag.mismatch_ID = mismatch.mismatch_ID \n",
    "WHERE (one_hum_multiple_prim = 1\n",
    "OR non_canonical_hum_spl = 1\n",
    "OR N_in_genomic = 1\n",
    "OR small_introns = 1\n",
    "OR frameshift = 1)\n",
    "AND conserved = 0\n",
    "AND repeats_prot = 0\n",
    "AND alignement_error = 0\n",
    "AND human_isoform_exist = 0\n",
    "\"\"\", conn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "61853    2641\n",
       "30611    1887\n",
       "9555     1091\n",
       "9595     1043\n",
       "60711     992\n",
       "9601      802\n",
       "9541      703\n",
       "9483      694\n",
       "9598      601\n",
       "9544      561\n",
       "Name: organism, dtype: int64"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(mismatch[\"organism\"].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "61853    1867\n",
       "30611    1145\n",
       "9595      566\n",
       "9555      408\n",
       "60711     379\n",
       "9601      339\n",
       "9541      269\n",
       "9483      257\n",
       "9598      184\n",
       "9544       32\n",
       "Name: organism, dtype: int64"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mismatch_mispredicted[\"organism\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5446\n",
      "4539\n"
     ]
    }
   ],
   "source": [
    "mismatch_to_correct = pd.read_sql_query(\n",
    "\"\"\"SELECT mismatch.*\n",
    "FROM mismatch_flag \n",
    "JOIN mismatch on mismatch_flag.mismatch_ID = mismatch.mismatch_ID \n",
    "WHERE (one_hum_multiple_prim = 1\n",
    "OR non_canonical_hum_spl = 1\n",
    "OR N_in_genomic = 1\n",
    "OR small_introns = 1\n",
    "OR frameshift = 1)\n",
    "AND conserved = 0\n",
    "AND repeats_prot = 0\n",
    "AND alignement_error = 0\n",
    "AND human_isoform_exist = 0\n",
    "\"\"\", conn)\n",
    "mismatch_to_correct\n",
    "aaa = mismatch_to_correct[\"prot_prim\"].tolist()\n",
    "print(len(aaa))\n",
    "print(len(list(set(aaa))))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.7.6 64-bit ('stage-env': conda)",
   "language": "python",
   "name": "python37664bitstageenvconda39052ae7dd75404f9bdb3226673ac6d7"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
