{
 "nbformat": 4,
 "nbformat_minor": 2,
 "metadata": {
  "language_info": {
   "name": "python",
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "version": "3.7.6-final"
  },
  "orig_nbformat": 2,
  "file_extension": ".py",
  "mimetype": "text/x-python",
  "name": "python",
  "npconvert_exporter": "python",
  "pygments_lexer": "ipython3",
  "version": 3,
  "kernelspec": {
   "name": "python37664bitstageenvconda39052ae7dd75404f9bdb3226673ac6d7",
   "display_name": "Python 3.7.6 64-bit ('stage-env': conda)"
  }
 },
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sqlite3\n",
    "import pandas as pd\n",
    "import os\n",
    "import sys\n",
    "from Bio.SubsMat import MatrixInfo as matlist\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "import math\n",
    "import numpy as np\n",
    "def fasta2List(pathFasta):\n",
    "    f = open(pathFasta, \"r\")\n",
    "    title = []\n",
    "    seq = []\n",
    "    seq_temp = []\n",
    "    for line in f:\n",
    "        if line[0] == \">\":\n",
    "            seq.append(''.join(seq_temp).replace(\"\\n\", \"\"))\n",
    "            title.append(line.replace(\"\\n\", \"\"))\n",
    "            seq_temp = []\n",
    "        else:\n",
    "            seq_temp.append(line)\n",
    "    seq.append(''.join(seq_temp).replace(\"\\n\", \"\"))\n",
    "    seq.pop(0)\n",
    "    dictionary = dict(zip(title, seq))\n",
    "    return dictionary\n",
    "\n",
    "conn = sqlite3.connect('../../mismatch_db.db')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "mismatch = pd.read_sql_query(\"SELECT * FROM mismatch\", conn)\n",
    "mismatch = mismatch.astype({\"exon_start_prim\": \"Int64\", \"exon_stop_prim\":\"Int64\", \"exon_start_hum\": \"Int64\", \"exon_stop_hum\":\"Int64\"})\n",
    "\n",
    "prot_seq_hum = pd.read_sql_query(\"\"\"\n",
    "SELECT mismatch.mismatch_ID, mismatch.prot_hum, protein.sequence\n",
    "FROM mismatch\n",
    "JOIN protein ON mismatch.prot_hum = protein.prot_ID\"\"\", conn)\n",
    "\n",
    "prot_seq_prim = pd.read_sql_query(\"\"\"\n",
    "SELECT mismatch.mismatch_ID, mismatch.prot_prim, protein.sequence\n",
    "FROM mismatch\n",
    "JOIN protein ON mismatch.prot_prim = protein.prot_ID\"\"\", conn)\n",
    "\n",
    "prim_exon_introns = pd.read_sql_query(\"\"\"\n",
    "SELECT mismatch_ID, mismatch.prot_prim, exon_intron_map.'type', exon_intron_map.number_elem, exon_intron_map.seq\n",
    "FROM mismatch\n",
    "JOIN protein ON mismatch.prot_prim = protein.prot_ID\n",
    "JOIN transcript ON protein.transcript_ID = transcript.transcript_ID\n",
    "JOIN exon_intron_map ON transcript.transcript_ID = exon_intron_map.transcript_ID\n",
    "\"\"\", conn)\n",
    "\n",
    "human_exon_introns = pd.read_sql_query(\"\"\"\n",
    "SELECT mismatch_ID, mismatch.prot_hum, exon_intron_map.'type', exon_intron_map.number_elem, exon_intron_map.seq\n",
    "FROM mismatch\n",
    "JOIN protein ON mismatch.prot_hum = protein.prot_ID\n",
    "JOIN transcript ON protein.transcript_ID = transcript.transcript_ID\n",
    "JOIN exon_intron_map ON transcript.transcript_ID = exon_intron_map.transcript_ID\n",
    "\"\"\", conn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": "728\n19012\n[45, 46, 87, 90, 91, 92, 112, 113, 115, 116, 117, 118, 256, 258, 260, 261, 263, 264, 265, 269, 270, 272, 273, 276, 278, 280, 281, 283, 284, 285, 289, 290, 292, 293, 296, 298, 300, 301, 303, 304, 307, 308, 309, 313, 314, 317, 318, 400, 402, 404, 405, 407, 408, 411, 412, 413, 417, 418, 421, 422, 427, 429, 431, 432, 434, 435, 436, 440, 441, 443, 444, 462, 463, 464, 465, 636, 638, 639, 640, 719, 722, 727, 728, 794, 797, 798, 799, 1635, 2038, 2040, 2114, 2149, 2152, 2153, 2311, 2312, 2410, 2411, 2412, 2413, 2414, 2415, 2487, 2488, 2492, 2494, 2500, 2509, 2556, 2557, 2559, 2561, 2592, 2594, 2597, 2598, 2599, 2600, 2601, 2722, 2723, 2725, 2727, 3023, 3024, 3025, 3026, 3027, 3028, 3030, 3031, 3032, 3033, 3034, 3035, 3067, 3161, 3164, 3169, 3171, 3468, 3469, 3470, 3471, 3472, 3556, 3557, 3558, 3560, 3561, 3584, 3601, 3603, 3605, 3606, 3607, 3866, 3867, 3868, 3870, 3871, 3872, 3942, 3943, 3946, 3947, 3949, 3950, 3951, 3954, 3955, 3967, 4222, 4223, 4224, 4226, 4229, 4384, 4385, 4386, 4387, 4388, 4407, 4468, 4469, 4470, 4472, 4575, 4577, 4579, 4580, 4798, 4800, 4802, 4805, 4840, 4968, 5160, 5161, 5163, 5164, 5165, 5166, 5167, 5268, 5312, 5330, 5331, 5332, 5371, 5373, 5374, 5514, 5516, 5517, 5518, 5520, 5521, 5524, 5526, 5527, 5528, 5529, 5531, 5637, 5638, 5639, 5640, 5732, 5733, 5734, 5735, 5737, 5738, 5739, 5743, 5744, 5746, 5747, 5807, 5820, 5821, 5822, 5823, 6009, 6010, 6011, 6012, 6013, 6456, 6457, 6458, 6459, 6460, 6461, 6462, 6463, 6464, 6465, 6466, 6467, 6602, 6603, 6604, 6605, 6700, 6705, 6706, 7262, 7263, 7264, 7265, 7266, 7267, 7317, 7389, 7393, 7394, 7396, 7713, 7714, 7715, 7716, 7754, 7755, 7756, 7757, 7859, 7861, 7863, 7864, 7865, 7904, 7907, 7925, 7926, 7931, 7933, 7936, 7937, 7938, 8007, 8012, 8013, 8014, 8019, 8023, 8033, 8166, 8254, 8381, 8382, 8385, 8387, 8467, 8470, 8471, 8472, 8473, 8551, 8553, 8555, 8567, 8643, 8678, 8679, 8680, 8681, 8713, 8714, 8715, 8716, 8717, 8737, 8738, 8740, 8742, 8743, 8774, 8837, 8839, 8841, 8844, 8846, 8847, 8903, 8906, 8907, 8909, 8910, 8911, 8936, 9031, 9109, 9359, 9468, 9469, 9470, 9471, 9530, 9531, 9532, 9535, 9536, 9553, 9554, 9555, 9556, 9557, 9558, 9559, 9560, 9561, 9562, 9563, 9564, 9565, 9673, 9675, 9678, 9798, 9799, 9800, 9801, 10042, 10089, 10091, 10102, 10103, 10104, 10106, 10107, 10242, 10243, 10270, 10280, 10337, 10338, 10396, 10397, 10398, 10399, 10400, 10431, 10432, 10433, 10434, 10435, 10436, 10437, 10488, 10499, 10574, 10611, 10614, 10628, 10629, 10630, 10631, 10812, 10813, 10814, 10815, 10817, 10818, 10819, 10820, 10821, 10822, 10824, 10825, 10826, 10827, 10828, 10830, 10831, 11051, 11081, 11082, 11084, 11085, 11086, 11196, 11197, 11198, 11240, 11242, 11243, 11244, 11377, 11485, 11730, 11770, 11772, 11773, 11807, 11835, 11949, 11950, 11951, 11952, 11953, 11954, 11957, 11959, 11961, 11962, 11963, 11964, 11965, 11966, 11986, 11998, 12000, 12004, 12005, 12007, 12013, 12014, 12015, 12016, 12017, 12160, 12208, 12212, 12213, 12214, 12382, 12386, 12387, 12388, 12452, 12528, 12529, 12530, 12532, 12533, 12534, 12535, 12536, 12540, 12541, 12543, 12545, 12633, 12655, 12656, 12659, 12660, 12662, 12663, 12664, 12665, 12666, 12667, 12669, 12670, 12671, 12672, 12673, 12872, 12874, 12875, 12876, 12985, 12986, 12987, 12988, 12989, 12990, 12991, 12992, 13188, 13189, 13190, 13191, 13192, 13229, 13309, 13310, 13311, 13313, 13314, 13667, 13668, 13669, 13670, 13675, 13676, 13677, 13678, 13679, 13680, 13697, 13701, 13702, 13703, 13704, 13706, 13708, 13786, 13787, 13790, 13791, 13801, 13913, 13914, 13934, 13935, 13937, 13938, 13939, 14023, 14025, 14027, 14028, 14262, 14263, 14264, 14265, 14266, 14267, 14268, 14361, 14362, 14363, 14364, 14365, 14369, 14559, 14622, 14623, 14624, 14628, 14632, 14772, 14843, 15020, 15065, 15116, 15117, 15118, 15119, 15120, 15121, 15122, 15254, 15261, 15412, 15713, 15715, 15797, 15798, 15799, 15800, 15801, 15869, 15894, 15895, 15896, 15897, 15899, 15901, 16021, 16022, 16096, 16097, 16120, 16121, 16122, 16123, 16124, 16133, 16134, 16136, 16137, 16169, 16270, 16271, 16272, 16273, 16303, 16304, 16307, 16308, 16375, 16481, 16482, 16483, 16484, 16485, 16486, 16809, 16896, 16897, 16898, 16899, 17009, 17010, 17011, 17012, 17100, 17101, 17107, 17372, 17373, 17374, 17375, 17565, 17566, 17567, 17568, 17911, 17912, 17913, 17914, 17915, 17916, 18087, 18091, 18092, 18093, 18152, 18154, 18168, 18169, 18171, 18172, 18173, 18174, 18175, 18176, 18177, 18178, 18179, 18180, 18310, 18311, 18318, 18319, 18353, 18460, 18493, 18565, 18566, 18567, 18568, 18569, 18614, 18616, 18618, 18622, 18830, 18882, 18883, 18884, 18885, 18886, 18935, 18936, 18937, 18939, 18991, 18992, 18993, 18994, 18995, 18996]\n"
    }
   ],
   "source": [
    "# Flagging des mismatch conservés V\n",
    "counter = 0\n",
    "total = 0\n",
    "index_conserved = []\n",
    "for index, row in mismatch.iloc[:,:].iterrows():\n",
    "    total+=1\n",
    "    conserved = 0\n",
    "\n",
    "    mySeq = fasta2List(\"../../data/raw/uniprot-sequence/\"+row[1]+\".id.fasta\")\n",
    "    mismtaching_seq = row[11]\n",
    "\n",
    "    for i, j in mySeq.items():\n",
    "        myAlign = pairwise2.align.localms(j, mismtaching_seq, 2, 0, -.5, -.1, one_alignment_only=True, score_only=True)\n",
    "        #print(format_alignment(*myAlign[0]))\n",
    "        if myAlign/len(mismtaching_seq) > 1.8:\n",
    "            conserved +=1\n",
    "\n",
    "    if conserved >= 4:\n",
    "        index_conserved.append(index)\n",
    "        #print(str(index)+'\\t'.join([str(x) for x in row]))\n",
    "        counter +=1\n",
    "print(counter)\n",
    "print(total)\n",
    "print(index_conserved)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": "160\n18714\n[45, 46, 462, 463, 464, 465, 1690, 2038, 2040, 2129, 2149, 2151, 2152, 2153, 2293, 2311, 2312, 3161, 3164, 3556, 3557, 3558, 3560, 3561, 3584, 3906, 3967, 4478, 4840, 5330, 5331, 5332, 5371, 5807, 6599, 6700, 6705, 6706, 6874, 7317, 7907, 8007, 8012, 8013, 8014, 8019, 8023, 8033, 8254, 8381, 8382, 8385, 8387, 8528, 8551, 8553, 8555, 8556, 8567, 8837, 8839, 8841, 8844, 8846, 8847, 9109, 9359, 9626, 9628, 10042, 10499, 10574, 11311, 11314, 11807, 12013, 12014, 12015, 12016, 12017, 12059, 12074, 12428, 12450, 12452, 12453, 12454, 12633, 12655, 12656, 12659, 12660, 12662, 12663, 12664, 12665, 12667, 12669, 12670, 12671, 12672, 12673, 13697, 13701, 13702, 13703, 13704, 13706, 13708, 13801, 13913, 13914, 13934, 13935, 13937, 13938, 13939, 14369, 14559, 14772, 15116, 15117, 15118, 15119, 15120, 15121, 15122, 15412, 15713, 15715, 16021, 16022, 16096, 16097, 16169, 16375, 17009, 17010, 17011, 17012, 17100, 17101, 17107, 17431, 17838, 18310, 18311, 18318, 18319, 18493, 18565, 18566, 18567, 18568, 18569, 18882, 18883, 18884, 18885, 18886]\n"
    }
   ],
   "source": [
    "# Flagging des erreurs d'alignement V\n",
    "counter = 0\n",
    "total = 0\n",
    "index_align_error = []\n",
    "for index, row in mismatch.iloc[:,:].iterrows():\n",
    "    try:\n",
    "        human_seq = prot_seq_hum.loc[prot_seq_hum[\"prot_hum\"]== row[1]].iloc[0,2]\n",
    "    except:\n",
    "        continue\n",
    "    total +=1\n",
    "    mismtaching_seq = row[11]\n",
    "    myAlign = pairwise2.align.localms(human_seq, mismtaching_seq, 2, 0, -.5, -.1, one_alignment_only=True, score_only=True)\n",
    "    if myAlign/len(mismtaching_seq) > 1.8:\n",
    "        index_align_error.append(index)\n",
    "        #print(str(index)+'\\t'.join([str(x) for x in row]))\n",
    "        counter += 1\n",
    "\n",
    "print(counter)\n",
    "print(total)\n",
    "print(index_align_error)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": "257\n18714\n[34, 35, 36, 37, 38, 39, 42, 43, 609, 884, 1000, 1001, 1002, 1385, 1391, 1614, 1618, 1619, 1620, 1621, 1622, 1623, 1625, 1626, 1627, 1631, 1639, 1746, 1755, 1757, 1758, 1759, 1761, 1847, 1848, 1849, 1872, 2061, 2065, 2256, 2301, 2346, 2359, 2365, 2367, 2368, 2370, 2372, 2373, 2424, 2564, 2565, 2567, 2568, 2573, 3160, 3552, 3878, 3879, 3880, 3881, 3882, 3883, 3885, 3886, 3887, 3888, 3889, 3890, 3891, 3893, 3896, 3897, 3898, 3900, 3901, 3903, 3906, 6150, 6586, 6592, 6699, 6700, 6702, 6703, 6704, 6705, 6706, 6924, 7113, 7332, 7539, 7542, 7882, 7884, 7885, 7886, 7904, 7907, 8021, 8023, 8105, 8106, 8107, 8108, 8109, 8110, 8111, 8112, 8113, 8254, 8386, 8391, 8392, 8393, 8394, 8395, 8398, 8400, 8401, 8403, 8404, 8405, 8406, 8407, 8410, 8412, 8414, 8415, 8416, 8417, 8418, 8419, 8551, 8558, 8560, 8561, 8562, 8563, 8565, 8566, 8569, 8837, 8839, 8841, 8844, 8846, 8847, 10488, 10491, 10495, 11256, 11675, 11676, 11677, 11678, 11679, 11680, 11681, 11682, 11683, 11684, 11689, 11690, 11691, 11692, 11694, 11695, 11696, 11698, 11727, 11931, 12049, 12051, 12052, 12058, 12072, 12074, 12096, 12097, 12098, 12099, 12100, 12217, 12225, 12315, 12316, 12319, 12320, 12329, 12331, 12334, 12339, 12450, 12657, 12661, 12662, 12663, 12664, 12665, 12667, 13399, 13766, 13916, 13926, 13935, 13938, 13939, 13981, 14369, 14808, 14809, 15170, 15171, 15181, 15183, 15184, 15364, 15989, 15990, 15994, 15995, 16013, 16021, 16098, 16099, 16100, 16102, 16168, 16169, 16987, 17010, 17110, 17707, 17709, 17710, 17711, 17712, 17714, 17715, 17718, 17723, 17726, 17804, 17828, 17837, 18314, 18315, 18428, 18430, 18488, 18491, 18542, 18544, 18882, 18883, 18884]\n"
    }
   ],
   "source": [
    "# Flagging des seq à repeats proteiq V\n",
    "counter = 0\n",
    "total = 0\n",
    "index_repeats = []\n",
    "for index, row in mismatch.iloc[:,:].iterrows():\n",
    "    try:\n",
    "        human_prot_Mismatch = prot_seq_hum.loc[prot_seq_hum[\"prot_hum\"]== row[1]].iloc[0,2]\n",
    "    except:\n",
    "        continue\n",
    "\n",
    "    total+=1\n",
    "    human_seq = row[12]\n",
    "    begin_seq = human_prot_Mismatch[:row[5]+1]\n",
    "    end_seq = human_prot_Mismatch[row[6]+2:]\n",
    "    human_prot_Mismatch = begin_seq + end_seq\n",
    "    myAlign = pairwise2.align.localms(human_prot_Mismatch, human_seq, 2, 0, -.5, -.1, one_alignment_only=True, score_only=True)\n",
    "    if myAlign/len(human_seq) > 1.8:\n",
    "        index_repeats.append(index)\n",
    "        counter += 1\n",
    "\n",
    "print(counter)\n",
    "print(total)\n",
    "print(index_repeats)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": "4723\n17470\n"
    }
   ],
   "source": [
    "#Flagging des seq à N \n",
    "\n",
    "counter = 0\n",
    "total = 0\n",
    "index_genom_n = []\n",
    "for index, row in mismatch.iloc[:,:].iterrows():\n",
    "    my_CDS = []\n",
    "    subset_exon_intron = prim_exon_introns.loc[prim_exon_introns[\"mismatch_ID\"] == index]\n",
    "    if isinstance(row[7], int) == False or isinstance(row[8], int) == False:\n",
    "        continue\n",
    "    try:\n",
    "        for n in range(row[7], row[8]+1):\n",
    "            my_exon = subset_exon_intron.loc[(subset_exon_intron[\"number_elem\"] == n) & (subset_exon_intron[\"type\"] == \"Exon\")]\n",
    "            my_intron = subset_exon_intron.loc[(subset_exon_intron[\"number_elem\"] == n) & (subset_exon_intron[\"type\"] == \"Intron\")]\n",
    "            try:\n",
    "                my_CDS.append(my_exon.iloc[0,4])\n",
    "                my_CDS.append(my_intron.iloc[0,4])\n",
    "            except:\n",
    "                pass\n",
    "        if \"N\" in ''.join(my_CDS):\n",
    "            index_genom_n.append(index)\n",
    "            counter +=1\n",
    "        total +=1\n",
    "    except:\n",
    "        pass\n",
    "\n",
    "print(counter)\n",
    "print(total)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": "470\n16061\n[54, 110, 137, 175, 232, 238, 323, 327, 353, 461, 534, 665, 680, 687, 731, 773, 774, 782, 783, 823, 937, 956, 975, 1022, 1026, 1029, 1112, 1117, 1118, 1143, 1149, 1251, 1294, 1320, 1329, 1352, 1353, 1393, 1419, 1420, 1422, 1428, 1488, 1494, 1548, 1554, 1621, 1768, 1813, 1854, 1973, 2028, 2035, 2042, 2149, 2422, 2433, 2439, 2536, 2722, 2723, 2725, 2726, 2727, 2747, 2756, 2768, 2775, 2800, 2815, 2821, 2930, 2979, 3053, 3079, 3126, 3130, 3163, 3193, 3201, 3220, 3309, 3326, 3344, 3360, 3466, 3476, 3493, 3644, 3645, 3665, 3738, 3763, 3775, 3885, 3937, 3998, 4003, 4023, 4085, 4112, 4236, 4243, 4259, 4265, 4285, 4315, 4360, 4428, 4534, 4605, 4691, 4787, 4902, 4918, 4960, 5080, 5094, 5144, 5176, 5177, 5178, 5214, 5254, 5256, 5344, 5394, 5400, 5406, 5416, 5431, 5447, 5461, 5589, 5598, 5599, 5650, 5653, 5659, 5660, 5666, 5672, 5673, 5688, 5708, 5892, 5909, 5947, 6185, 6260, 6290, 6382, 6468, 6475, 6485, 6496, 6498, 6556, 6737, 6775, 6849, 6856, 6951, 7031, 7083, 7111, 7112, 7204, 7209, 7214, 7219, 7257, 7312, 7472, 7539, 7540, 7571, 7835, 7923, 7939, 7949, 7950, 7951, 8018, 8066, 8085, 8138, 8232, 8286, 8303, 8310, 8312, 8399, 8405, 8413, 8536, 8682, 8713, 8714, 8716, 8717, 8766, 8780, 8834, 8919, 8966, 8967, 8973, 8986, 9010, 9047, 9051, 9052, 9241, 9284, 9285, 9318, 9337, 9395, 9407, 9452, 9460, 9461, 9538, 9593, 9607, 9655, 9665, 9666, 9682, 9693, 9744, 9745, 9754, 9761, 9770, 9816, 9906, 10048, 10050, 10067, 10089, 10091, 10092, 10124, 10158, 10162, 10197, 10326, 10367, 10377, 10459, 10513, 10537, 10669, 10673, 10705, 10732, 10772, 10816, 10859, 10874, 10883, 10916, 10945, 10998, 11026, 11034, 11096, 11142, 11191, 11312, 11352, 11390, 11396, 11425, 11440, 11447, 11522, 11540, 11584, 11612, 11613, 11629, 11631, 11669, 11679, 11747, 11882, 11897, 11934, 12083, 12125, 12178, 12225, 12231, 12283, 12290, 12320, 12406, 12415, 12456, 12516, 12528, 12529, 12530, 12534, 12535, 12623, 12624, 12657, 12668, 12705, 12813, 12825, 12911, 12951, 12963, 12968, 12995, 13023, 13031, 13039, 13059, 13180, 13309, 13314, 13384, 13474, 13495, 13528, 13533, 13534, 13563, 13576, 13652, 13654, 13797, 13806, 13928, 13966, 13988, 13995, 14013, 14064, 14066, 14203, 14258, 14285, 14368, 14406, 14410, 14411, 14492, 14610, 14683, 14761, 14778, 14794, 14795, 14796, 14801, 14805, 14837, 14843, 14988, 15005, 15006, 15008, 15068, 15071, 15088, 15242, 15385, 15439, 15440, 15456, 15494, 15629, 15638, 15668, 15699, 15747, 15779, 15781, 15785, 15803, 15806, 15847, 15875, 15919, 15987, 15990, 15992, 15993, 15994, 16007, 16023, 16061, 16165, 16203, 16262, 16299, 16312, 16314, 16319, 16331, 16424, 16436, 16503, 16668, 16701, 16875, 16889, 16997, 17023, 17082, 17105, 17115, 17138, 17151, 17159, 17167, 17171, 17226, 17233, 17408, 17409, 17410, 17492, 17534, 17548, 17562, 17563, 17639, 17748, 17863, 17944, 18072, 18101, 18115, 18183, 18196, 18213, 18242, 18302, 18320, 18345, 18346, 18351, 18388, 18401, 18419, 18503, 18504, 18511, 18606, 18614, 18616, 18618, 18622, 18790, 18826, 18946, 18965, 18991, 18992, 18994, 18995, 18996]\n"
    }
   ],
   "source": [
    "# Flagging des 1 exon human vs multiple primate V\n",
    "counter = 0\n",
    "total = 0\n",
    "index_multiple_exon = []\n",
    "for index, row in mismatch.iloc[:, :].iterrows():\n",
    "    if isinstance(row[7], int) == False or isinstance(row[9], int) == False:\n",
    "        continue\n",
    "    total +=1\n",
    "    if (int(row[9])-int(row[10]) == 0) and (row[8]-row[7] >= 2):\n",
    "        #print(str(index)+'\\t'.join([str(x) for x in row]))\n",
    "        index_multiple_exon.append(index)\n",
    "        counter += 1\n",
    "\n",
    "print(counter)\n",
    "print(total)\n",
    "print(index_multiple_exon)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": "Canonical (GT/AG) 99.06727804188617\nNon canonical 0.9327219581138378\n"
    }
   ],
   "source": [
    "# Stat sur les sites introniques Humains\n",
    "\n",
    "canonical = 0\n",
    "non_canonical = 0\n",
    "human_intron_seq = human_exon_introns.loc[human_exon_introns[\"type\"]==\"Intron\"]\n",
    "for index, row in human_intron_seq.iloc[:, [0,4]].iterrows():\n",
    "    if row[1][:2] == \"GT\" and row[1][-2:] == \"AG\":\n",
    "        canonical += 1\n",
    "    #elif row[6][:2] == \"GC\" and row[6][-2:] == \"AG\":\n",
    "    #    canonical += 1\n",
    "    #elif row[6][:2] == \"AT\" and row[6][-2:] == \"AC\":\n",
    "    #    canonical += 1\n",
    "    else:\n",
    "        non_canonical += 1\n",
    "print(\"Canonical (GT/AG) \" + str(canonical/(canonical+non_canonical)*100))\n",
    "print(\"Non canonical \" + str(non_canonical/(canonical+non_canonical)*100))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>mismatch_ID</th>\n      <th>prot_hum</th>\n      <th>prot_prim</th>\n      <th>pos_start_prim</th>\n      <th>pos_stop_prim</th>\n      <th>pos_start_hum</th>\n      <th>pos_stop_hum</th>\n      <th>exon_start_prim</th>\n      <th>exon_stop_prim</th>\n      <th>exon_start_hum</th>\n      <th>exon_stop_hum</th>\n      <th>seq_prim</th>\n      <th>seq_hum</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>0</td>\n      <td>A0A044PY82</td>\n      <td>H0WKS4</td>\n      <td>1</td>\n      <td>55</td>\n      <td>0</td>\n      <td>50</td>\n      <td>1</td>\n      <td>3</td>\n      <td>1</td>\n      <td>2</td>\n      <td>AGNPAQPSPGRESLARSKAAPGAPSGLSCAQQGCRGRRPALPLGGR...</td>\n      <td>MAPQRAGPGLGKPSSKKSRALRGGPGVPRARQRSPRAAAELNTEVP...</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>1</td>\n      <td>A0A044PY82</td>\n      <td>H0WKS4</td>\n      <td>108</td>\n      <td>119</td>\n      <td>110</td>\n      <td>122</td>\n      <td>3</td>\n      <td>3</td>\n      <td>2</td>\n      <td>2</td>\n      <td>KISSGSSLETPQ</td>\n      <td>PKTSSSRSGSEAP</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>2</td>\n      <td>A0A044PY82</td>\n      <td>G3QIT3</td>\n      <td>92</td>\n      <td>111</td>\n      <td>91</td>\n      <td>119</td>\n      <td>2</td>\n      <td>3</td>\n      <td>2</td>\n      <td>2</td>\n      <td>GPSDRSAKGGGPGAASPSSP</td>\n      <td>PGPSEAQRAEVQGQHPPPGPKTSSSRSGS</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>3</td>\n      <td>A0A075B6I0</td>\n      <td>A0A0D9RMQ9</td>\n      <td>74</td>\n      <td>97</td>\n      <td>78</td>\n      <td>117</td>\n      <td>2</td>\n      <td>2</td>\n      <td>2</td>\n      <td>2</td>\n      <td>CPSGVPESLLWLHPWEQSCPHHHG</td>\n      <td>TRSSGVPDRFSGSILGNKAALTITGAQADDESDYYCVLYM</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>4</td>\n      <td>A0A075B6I3</td>\n      <td>A0A096NSB3</td>\n      <td>36</td>\n      <td>51</td>\n      <td>35</td>\n      <td>50</td>\n      <td>2</td>\n      <td>2</td>\n      <td>2</td>\n      <td>2</td>\n      <td>ASLTCTFSGGINVAGY</td>\n      <td>TARLPCTLSSDLSVGG</td>\n    </tr>\n    <tr>\n      <th>...</th>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n    </tr>\n    <tr>\n      <th>19007</th>\n      <td>19007</td>\n      <td>Q6UXU4</td>\n      <td>F7ILG0</td>\n      <td>4</td>\n      <td>37</td>\n      <td>0</td>\n      <td>30</td>\n      <td>2</td>\n      <td>2</td>\n      <td>1</td>\n      <td>1</td>\n      <td>ELAKAFSSQRTRLSAILSMLSLSFSTISLFSNYW</td>\n      <td>MKTSRRGRALLAVALNLLALLFATTAFLTTH</td>\n    </tr>\n    <tr>\n      <th>19008</th>\n      <td>19008</td>\n      <td>Q6UXU4</td>\n      <td>F7ILG0</td>\n      <td>44</td>\n      <td>77</td>\n      <td>37</td>\n      <td>83</td>\n      <td>2</td>\n      <td>2</td>\n      <td>1</td>\n      <td>1</td>\n      <td>VPKPLCEKDLAAKCFDMPVSLDGDVANTSTQEVV</td>\n      <td>RVPKPGCGQGGRANCPNSGANATANGTAAPAAAAAAATASGNGPPGG</td>\n    </tr>\n    <tr>\n      <th>19009</th>\n      <td>19009</td>\n      <td>Q6UXU4</td>\n      <td>F7ILG0</td>\n      <td>115</td>\n      <td>245</td>\n      <td>121</td>\n      <td>273</td>\n      <td>3</td>\n      <td>5</td>\n      <td>2</td>\n      <td>5</td>\n      <td>FIELTPPTERGEKGLLEFATLQGPCHPTLRFGGKQLMEKASLPYPP...</td>\n      <td>SFIDLAPASEKGVLWLSVVSEVLYILLLVVGFSLMCLELFHSSNVI...</td>\n    </tr>\n    <tr>\n      <th>19010</th>\n      <td>19010</td>\n      <td>Q6UXU4</td>\n      <td>F7ILG0</td>\n      <td>247</td>\n      <td>286</td>\n      <td>276</td>\n      <td>327</td>\n      <td>5</td>\n      <td>6</td>\n      <td>5</td>\n      <td>7</td>\n      <td>LELRLGLLHGLAFLHLLHGVGGHHLQYVHQDGAGVQVQA</td>\n      <td>RMEKRDGSEEDFHLDCRHERYPARHQPHMADSWPRSSAQEAPELNR...</td>\n    </tr>\n    <tr>\n      <th>19011</th>\n      <td>19011</td>\n      <td>Q6UXU6</td>\n      <td>H0XV48</td>\n      <td>80</td>\n      <td>95</td>\n      <td>79</td>\n      <td>96</td>\n      <td>4</td>\n      <td>4</td>\n      <td>5</td>\n      <td>5</td>\n      <td>RSCIQPEQRNLQESSE</td>\n      <td>CRNCREPEPDSPVDCRGP</td>\n    </tr>\n  </tbody>\n</table>\n<p>19012 rows × 13 columns</p>\n</div>",
      "text/plain": "       mismatch_ID    prot_hum   prot_prim  pos_start_prim  pos_stop_prim  \\\n0                0  A0A044PY82      H0WKS4               1             55   \n1                1  A0A044PY82      H0WKS4             108            119   \n2                2  A0A044PY82      G3QIT3              92            111   \n3                3  A0A075B6I0  A0A0D9RMQ9              74             97   \n4                4  A0A075B6I3  A0A096NSB3              36             51   \n...            ...         ...         ...             ...            ...   \n19007        19007      Q6UXU4      F7ILG0               4             37   \n19008        19008      Q6UXU4      F7ILG0              44             77   \n19009        19009      Q6UXU4      F7ILG0             115            245   \n19010        19010      Q6UXU4      F7ILG0             247            286   \n19011        19011      Q6UXU6      H0XV48              80             95   \n\n       pos_start_hum  pos_stop_hum  exon_start_prim  exon_stop_prim  \\\n0                  0            50                1               3   \n1                110           122                3               3   \n2                 91           119                2               3   \n3                 78           117                2               2   \n4                 35            50                2               2   \n...              ...           ...              ...             ...   \n19007              0            30                2               2   \n19008             37            83                2               2   \n19009            121           273                3               5   \n19010            276           327                5               6   \n19011             79            96                4               4   \n\n       exon_start_hum  exon_stop_hum  \\\n0                   1              2   \n1                   2              2   \n2                   2              2   \n3                   2              2   \n4                   2              2   \n...               ...            ...   \n19007               1              1   \n19008               1              1   \n19009               2              5   \n19010               5              7   \n19011               5              5   \n\n                                                seq_prim  \\\n0      AGNPAQPSPGRESLARSKAAPGAPSGLSCAQQGCRGRRPALPLGGR...   \n1                                           KISSGSSLETPQ   \n2                                   GPSDRSAKGGGPGAASPSSP   \n3                               CPSGVPESLLWLHPWEQSCPHHHG   \n4                                       ASLTCTFSGGINVAGY   \n...                                                  ...   \n19007                 ELAKAFSSQRTRLSAILSMLSLSFSTISLFSNYW   \n19008                 VPKPLCEKDLAAKCFDMPVSLDGDVANTSTQEVV   \n19009  FIELTPPTERGEKGLLEFATLQGPCHPTLRFGGKQLMEKASLPYPP...   \n19010            LELRLGLLHGLAFLHLLHGVGGHHLQYVHQDGAGVQVQA   \n19011                                   RSCIQPEQRNLQESSE   \n\n                                                 seq_hum  \n0      MAPQRAGPGLGKPSSKKSRALRGGPGVPRARQRSPRAAAELNTEVP...  \n1                                          PKTSSSRSGSEAP  \n2                          PGPSEAQRAEVQGQHPPPGPKTSSSRSGS  \n3               TRSSGVPDRFSGSILGNKAALTITGAQADDESDYYCVLYM  \n4                                       TARLPCTLSSDLSVGG  \n...                                                  ...  \n19007                    MKTSRRGRALLAVALNLLALLFATTAFLTTH  \n19008    RVPKPGCGQGGRANCPNSGANATANGTAAPAAAAAAATASGNGPPGG  \n19009  SFIDLAPASEKGVLWLSVVSEVLYILLLVVGFSLMCLELFHSSNVI...  \n19010  RMEKRDGSEEDFHLDCRHERYPARHQPHMADSWPRSSAQEAPELNR...  \n19011                                 CRNCREPEPDSPVDCRGP  \n\n[19012 rows x 13 columns]"
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mismatch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": "414\n16982\n[63, 67, 97, 102, 105, 124, 125, 161, 162, 163, 164, 165, 205, 223, 322, 324, 377, 389, 818, 820, 821, 846, 892, 893, 1107, 1108, 1114, 1116, 1194, 1195, 1196, 1231, 1260, 1402, 1766, 1773, 1779, 2033, 2034, 2039, 2041, 2052, 2053, 2054, 2058, 2059, 2060, 2160, 2297, 2298, 2301, 2401, 2454, 2455, 2459, 2460, 2461, 2462, 2463, 2464, 2465, 2466, 2467, 2468, 2555, 2556, 2557, 2559, 2561, 2637, 2715, 2788, 2903, 2904, 2943, 2946, 3006, 3009, 3064, 3084, 3187, 3206, 3308, 3415, 3421, 3422, 3423, 3424, 3425, 3426, 3428, 3488, 3542, 3627, 3628, 3704, 3728, 3833, 4006, 4280, 4281, 4374, 4444, 4519, 4525, 4526, 4597, 4623, 4681, 4682, 4684, 4696, 4699, 4703, 4707, 4724, 4727, 4728, 4729, 4754, 4797, 4828, 4876, 4878, 4887, 4888, 5004, 5076, 5088, 5199, 5208, 5394, 5397, 5398, 5434, 5456, 5461, 5462, 5533, 5549, 5633, 5635, 5766, 5801, 5804, 5813, 5814, 5815, 5816, 5818, 5819, 5830, 5831, 5832, 5898, 5899, 5925, 6401, 6503, 6557, 6567, 6613, 6695, 6945, 6946, 7101, 7103, 7104, 7106, 7108, 7109, 7220, 7388, 7424, 7499, 7509, 7510, 7511, 7559, 7560, 7576, 7712, 7748, 7751, 7765, 7784, 7850, 7983, 7984, 8017, 8018, 8019, 8020, 8299, 8343, 8344, 8422, 8447, 8497, 8622, 8929, 9048, 9049, 9099, 9220, 9357, 9398, 9410, 9411, 9485, 9624, 9739, 9771, 9772, 9774, 9827, 9866, 9868, 9869, 9885, 9887, 9947, 10328, 10391, 10394, 10395, 10424, 10425, 10451, 10452, 10454, 10467, 10468, 10469, 10500, 10501, 10561, 10582, 10583, 10585, 10663, 10777, 10796, 10968, 11416, 11456, 11457, 11721, 11722, 11723, 11857, 11932, 11933, 11968, 11970, 11971, 11997, 12225, 12226, 12228, 12229, 12492, 12561, 12691, 12692, 12695, 12704, 12706, 12707, 12773, 12804, 12805, 12806, 12824, 13050, 13110, 13134, 13187, 13206, 13208, 13209, 13210, 13303, 13304, 13327, 13349, 13414, 13443, 13494, 13497, 13561, 13584, 13602, 13684, 13724, 13776, 13793, 13797, 13827, 13828, 13836, 13837, 13838, 13839, 13875, 13884, 13885, 13964, 13973, 14141, 14248, 14316, 14358, 14359, 14386, 14431, 14488, 14495, 14537, 14738, 14745, 14798, 14867, 15062, 15076, 15077, 15101, 15102, 15123, 15127, 15267, 15287, 15288, 15289, 15290, 15314, 15315, 15316, 15317, 15318, 15319, 15320, 15437, 15438, 15475, 15499, 15547, 15624, 15650, 15672, 15690, 15723, 15904, 15905, 15906, 16038, 16039, 16040, 16068, 16069, 16241, 16390, 16391, 16418, 16419, 16649, 16749, 16902, 17063, 17067, 17152, 17153, 17174, 17218, 17310, 17311, 17391, 17564, 17911, 17912, 17913, 17914, 17915, 17916, 18138, 18139, 18286, 18309, 18326, 18327, 18329, 18331, 18354, 18356, 18357, 18358, 18400, 18403, 18445, 18546, 18606, 18612, 18624, 18697, 18717, 18725, 18728, 18730, 18731, 18781, 18930, 18943, 18981, 18983]\nCPU times: user 12min 58s, sys: 1.18 s, total: 12min 59s\nWall time: 12min 59s\n"
    }
   ],
   "source": [
    "%%time\n",
    "# Flagging des mismatch ayant un site humain de splicing non canonique V\n",
    "counter = 0\n",
    "total = 0\n",
    "index_non_canon = []\n",
    "for index, row in mismatch.iloc[:, :].iterrows():\n",
    "    subset_intron = human_exon_introns.loc[(human_exon_introns['mismatch_ID'] == row[0]) & (human_exon_introns['type'] == \"Intron\")]\n",
    "    if subset_intron.empty or row[9] == \"ERROR\":\n",
    "        continue\n",
    "    total += 1\n",
    "    for i in range(int(row[9]),int(row[10])+2):\n",
    "        intron = subset_intron.loc[subset_intron[\"number_elem\"] == (i-1)]\n",
    "        if intron.empty:\n",
    "            continue\n",
    "        intron = intron.iloc[0,:].to_list()\n",
    "        #if  (intron[6][:2] == \"GT\" and intron[6][-2:] == \"AG\") \\\n",
    "        #    or (intron[6][:2] == \"GC\" and intron[6][-2:] == \"AG\") \\\n",
    "        #    or (intron[6][:2] == \"AT\" and intron[6][-2:] == \"AC\") and (len(intron[6]) > 30):\n",
    "        if  (intron[4][:2] == \"GT\" and intron[4][-2:] == \"AG\") and (len(intron[4]) > 30):\n",
    "            pass\n",
    "        else:\n",
    "            index_non_canon.append(index)\n",
    "            counter +=1\n",
    "            #print(str(index)+'\\t'.join([str(x) for x in row]))\n",
    "            break        \n",
    "\n",
    "print(counter)\n",
    "print(total)\n",
    "print(index_non_canon)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": "903\n16936\n[21, 52, 67, 84, 89, 114, 126, 155, 254, 257, 277, 297, 341, 342, 343, 401, 428, 451, 461, 534, 555, 564, 571, 629, 684, 686, 687, 710, 726, 734, 735, 822, 823, 860, 868, 881, 911, 912, 913, 934, 935, 936, 946, 956, 957, 977, 1071, 1099, 1152, 1169, 1224, 1225, 1327, 1329, 1353, 1368, 1371, 1393, 1397, 1420, 1445, 1449, 1510, 1540, 1541, 1542, 1572, 1573, 1645, 1654, 1690, 1700, 1768, 1769, 1772, 1776, 1797, 1809, 1810, 1825, 1854, 1855, 1859, 1862, 2035, 2042, 2094, 2129, 2225, 2266, 2270, 2310, 2323, 2327, 2401, 2462, 2529, 2530, 2531, 2569, 2607, 2608, 2609, 2610, 2620, 2621, 2622, 2630, 2631, 2632, 2665, 2671, 2677, 2691, 2693, 2694, 2705, 2717, 2718, 2732, 2765, 2774, 2791, 2825, 2862, 2900, 2947, 2950, 2974, 2975, 2976, 3001, 3006, 3022, 3068, 3069, 3077, 3079, 3091, 3100, 3126, 3134, 3153, 3160, 3162, 3173, 3203, 3218, 3229, 3241, 3286, 3308, 3320, 3321, 3322, 3326, 3327, 3358, 3367, 3375, 3376, 3434, 3466, 3474, 3476, 3484, 3508, 3527, 3538, 3577, 3579, 3596, 3619, 3649, 3667, 3696, 3737, 3739, 3763, 3764, 3808, 3817, 3823, 3840, 3861, 3862, 3865, 3931, 3965, 4023, 4030, 4034, 4059, 4090, 4106, 4114, 4237, 4295, 4314, 4345, 4353, 4365, 4373, 4403, 4408, 4434, 4454, 4471, 4480, 4553, 4563, 4609, 4629, 4673, 4693, 4702, 4706, 4750, 4764, 4782, 4786, 4796, 4818, 4822, 4827, 4831, 4858, 4875, 4909, 4918, 4931, 4935, 4947, 4948, 4955, 4958, 5064, 5118, 5124, 5128, 5142, 5244, 5261, 5270, 5282, 5287, 5333, 5335, 5339, 5340, 5347, 5400, 5431, 5438, 5439, 5440, 5441, 5447, 5461, 5483, 5515, 5591, 5599, 5601, 5602, 5649, 5652, 5655, 5666, 5731, 5757, 5772, 5794, 5796, 5799, 5824, 5826, 5842, 5845, 5877, 5882, 5883, 5886, 5887, 5892, 5928, 5953, 5956, 5958, 5961, 6006, 6024, 6106, 6111, 6156, 6185, 6186, 6290, 6294, 6300, 6312, 6345, 6352, 6422, 6445, 6468, 6475, 6538, 6542, 6569, 6571, 6572, 6595, 6612, 6641, 6674, 6675, 6679, 6704, 6770, 6771, 6829, 6835, 6859, 6863, 6874, 6875, 6943, 7064, 7110, 7187, 7188, 7189, 7223, 7224, 7312, 7345, 7371, 7426, 7489, 7490, 7492, 7519, 7538, 7541, 7589, 7601, 7624, 7630, 7644, 7681, 7708, 7724, 7785, 7787, 7910, 7915, 7935, 7945, 7960, 8001, 8006, 8051, 8073, 8081, 8085, 8118, 8123, 8125, 8138, 8140, 8160, 8174, 8175, 8176, 8232, 8238, 8242, 8252, 8313, 8329, 8332, 8364, 8371, 8392, 8396, 8399, 8403, 8405, 8413, 8419, 8432, 8440, 8451, 8452, 8469, 8475, 8519, 8547, 8558, 8572, 8574, 8699, 8732, 8774, 8807, 8836, 8840, 8874, 8883, 8950, 8954, 8979, 8994, 9007, 9010, 9032, 9052, 9060, 9077, 9103, 9129, 9170, 9181, 9196, 9240, 9257, 9262, 9284, 9285, 9318, 9327, 9329, 9332, 9333, 9357, 9365, 9376, 9378, 9379, 9395, 9458, 9461, 9485, 9500, 9503, 9534, 9542, 9577, 9605, 9615, 9647, 9649, 9660, 9681, 9682, 9720, 9729, 9741, 9758, 9761, 9773, 9824, 9829, 9849, 9861, 9874, 9899, 9921, 9926, 9944, 9947, 9949, 10020, 10023, 10058, 10067, 10092, 10117, 10121, 10122, 10163, 10168, 10192, 10193, 10206, 10250, 10257, 10278, 10290, 10318, 10326, 10358, 10376, 10378, 10386, 10402, 10404, 10499, 10513, 10523, 10540, 10547, 10555, 10556, 10566, 10570, 10573, 10581, 10648, 10669, 10673, 10676, 10687, 10702, 10739, 10765, 10777, 10787, 10802, 10829, 10898, 10917, 10945, 10946, 10976, 11060, 11081, 11082, 11084, 11085, 11086, 11118, 11120, 11127, 11137, 11157, 11198, 11241, 11253, 11262, 11264, 11280, 11336, 11352, 11361, 11396, 11400, 11408, 11416, 11425, 11435, 11438, 11439, 11476, 11481, 11495, 11524, 11536, 11542, 11577, 11578, 11579, 11580, 11582, 11584, 11599, 11616, 11620, 11622, 11629, 11631, 11635, 11638, 11668, 11669, 11728, 11731, 11737, 11760, 11774, 11803, 11809, 11826, 11881, 11883, 11917, 11946, 11999, 12041, 12081, 12083, 12084, 12101, 12123, 12152, 12178, 12179, 12190, 12306, 12309, 12374, 12418, 12441, 12444, 12456, 12484, 12491, 12492, 12495, 12515, 12516, 12520, 12598, 12615, 12624, 12641, 12658, 12661, 12668, 12678, 12705, 12777, 12789, 12794, 12811, 12824, 12834, 12838, 12859, 12864, 12893, 12947, 12951, 13011, 13037, 13050, 13054, 13166, 13192, 13215, 13217, 13226, 13227, 13228, 13285, 13331, 13342, 13357, 13359, 13414, 13432, 13455, 13480, 13484, 13494, 13514, 13528, 13533, 13560, 13568, 13577, 13590, 13599, 13641, 13648, 13652, 13661, 13696, 13710, 13711, 13766, 13799, 13846, 13896, 13960, 13966, 13975, 13992, 14029, 14030, 14051, 14055, 14064, 14067, 14072, 14116, 14125, 14139, 14147, 14149, 14153, 14177, 14178, 14200, 14201, 14203, 14217, 14230, 14241, 14242, 14352, 14360, 14371, 14414, 14438, 14483, 14485, 14486, 14487, 14489, 14511, 14512, 14542, 14566, 14580, 14610, 14615, 14643, 14651, 14654, 14662, 14676, 14703, 14783, 14799, 14801, 14837, 14854, 14881, 14906, 14924, 14989, 15005, 15008, 15013, 15019, 15040, 15048, 15067, 15088, 15098, 15144, 15151, 15155, 15244, 15268, 15307, 15330, 15332, 15333, 15365, 15436, 15440, 15505, 15507, 15519, 15530, 15531, 15552, 15556, 15596, 15627, 15637, 15647, 15664, 15671, 15672, 15682, 15699, 15700, 15701, 15716, 15777, 15780, 15781, 15802, 15846, 15853, 15865, 15867, 15904, 15905, 16068, 16071, 16082, 16118, 16175, 16203, 16249, 16262, 16291, 16314, 16319, 16331, 16334, 16357, 16388, 16494, 16504, 16569, 16621, 16661, 16668, 16709, 16728, 16729, 16763, 16801, 16803, 16806, 16811, 16813, 16830, 16844, 16881, 16910, 16918, 16938, 16958, 16963, 17025, 17027, 17105, 17138, 17139, 17167, 17221, 17230, 17233, 17235, 17248, 17306, 17318, 17464, 17480, 17525, 17542, 17600, 17616, 17671, 17693, 17728, 17736, 17749, 17878, 17894, 17901, 17947, 17965, 17999, 18000, 18014, 18028, 18032, 18072, 18075, 18099, 18148, 18167, 18182, 18213, 18214, 18215, 18226, 18242, 18245, 18254, 18283, 18284, 18285, 18286, 18293, 18349, 18351, 18375, 18382, 18387, 18404, 18425, 18484, 18486, 18491, 18510, 18521, 18555, 18564, 18575, 18611, 18625, 18682, 18714, 18732, 18794, 18808, 18858, 18909, 18913, 18947, 18955, 18965, 18972, 18978]\n"
    }
   ],
   "source": [
    "# Flagging mismatch with too small introns V\n",
    "counter = 0\n",
    "total = 0\n",
    "index_intron_small = []\n",
    "for index, row in mismatch.iloc[:, :].iterrows():\n",
    "    subset_intron = prim_exon_introns.loc[(prim_exon_introns['mismatch_ID'] == row[0]) & (prim_exon_introns['type'] == \"Intron\")]\n",
    "    if subset_intron.empty or row[7] == \"ERROR\":\n",
    "        continue\n",
    "    total += 1\n",
    "    for i in range(int(row[7]),int(row[8])):\n",
    "        intron = subset_intron.loc[subset_intron[\"number_elem\"] == i]\n",
    "        intron = intron.iloc[0,:].to_list()\n",
    "        if len(intron[4]) <= 29:\n",
    "            index_intron_small.append(index)\n",
    "            counter +=1\n",
    "            #print(str(index)+'\\t'.join([str(x) for x in row]))\n",
    "            break\n",
    "\n",
    "print(counter)\n",
    "print(total)\n",
    "print(index_intron_small)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "column_names = [\"mismatch_ID\", \"conserved\", \"one_hum_multiple_prim\", \"non_canonical_hum_spl\", \"N_in_genomic\", \"small_introns\", \"repeats_prot\", \"alignement_error\", \"human_isoform_exist\"]\n",
    "df = pd.DataFrame(columns = column_names)\n",
    "df[\"mismatch_ID\"] = mismatch[\"mismatch_ID\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['conserved'] = np.where(df[\"mismatch_ID\"].isin(index_conserved), 1, 0)\n",
    "df['one_hum_multiple_prim'] = np.where(df[\"mismatch_ID\"].isin(index_multiple_exon), 1, 0)\n",
    "df['non_canonical_hum_spl'] = np.where(df[\"mismatch_ID\"].isin(index_non_canon), 1, 0)\n",
    "df['N_in_genomic'] = np.where(df[\"mismatch_ID\"].isin(index_genom_n), 1, 0)\n",
    "df['small_introns'] = np.where(df[\"mismatch_ID\"].isin(index_intron_small), 1, 0)\n",
    "df['repeats_prot'] = np.where(df[\"mismatch_ID\"].isin(index_repeats), 1, 0)\n",
    "df['alignement_error'] = np.where(df[\"mismatch_ID\"].isin(index_align_error), 1, 0)\n",
    "#df['human_isoform_exist'] = np.where(df[\"mismatch_ID\"].isin(index_conserved), 1, 0)\n",
    "#df.to_sql(con=conn, name='mismatch_flag', index=False, if_exists=\"append\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "tata = pd.read_sql_query(\n",
    "\"\"\"SELECT * FROM mismatch_flag \n",
    "WHERE conserved = 0 \n",
    "AND one_hum_multiple_prim = 0 \n",
    "AND non_canonical_hum_spl = 0\n",
    "AND N_in_genomic = 0\n",
    "AND small_introns = 0\n",
    "AND repeats_prot = 0\n",
    "AND alignement_error = 0\n",
    "\"\"\", conn)\n",
    "tata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>mismatch_ID</th>\n      <th>prot_hum</th>\n      <th>prot_prim</th>\n      <th>pos_start_prim</th>\n      <th>pos_stop_prim</th>\n      <th>pos_start_hum</th>\n      <th>pos_stop_hum</th>\n      <th>exon_start_prim</th>\n      <th>exon_stop_prim</th>\n      <th>exon_start_hum</th>\n      <th>exon_stop_hum</th>\n      <th>seq_prim</th>\n      <th>seq_hum</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>0</td>\n      <td>A0A044PY82</td>\n      <td>H0WKS4</td>\n      <td>1</td>\n      <td>55</td>\n      <td>0</td>\n      <td>50</td>\n      <td>1</td>\n      <td>3</td>\n      <td>1</td>\n      <td>2</td>\n      <td>AGNPAQPSPGRESLARSKAAPGAPSGLSCAQQGCRGRRPALPLGGR...</td>\n      <td>MAPQRAGPGLGKPSSKKSRALRGGPGVPRARQRSPRAAAELNTEVP...</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>1</td>\n      <td>A0A044PY82</td>\n      <td>H0WKS4</td>\n      <td>108</td>\n      <td>119</td>\n      <td>110</td>\n      <td>122</td>\n      <td>3</td>\n      <td>3</td>\n      <td>2</td>\n      <td>2</td>\n      <td>KISSGSSLETPQ</td>\n      <td>PKTSSSRSGSEAP</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>2</td>\n      <td>A0A044PY82</td>\n      <td>G3QIT3</td>\n      <td>92</td>\n      <td>111</td>\n      <td>91</td>\n      <td>119</td>\n      <td>2</td>\n      <td>3</td>\n      <td>2</td>\n      <td>2</td>\n      <td>GPSDRSAKGGGPGAASPSSP</td>\n      <td>PGPSEAQRAEVQGQHPPPGPKTSSSRSGS</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>3</td>\n      <td>A0A075B6I0</td>\n      <td>A0A0D9RMQ9</td>\n      <td>74</td>\n      <td>97</td>\n      <td>78</td>\n      <td>117</td>\n      <td>2</td>\n      <td>2</td>\n      <td>2</td>\n      <td>2</td>\n      <td>CPSGVPESLLWLHPWEQSCPHHHG</td>\n      <td>TRSSGVPDRFSGSILGNKAALTITGAQADDESDYYCVLYM</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>4</td>\n      <td>A0A075B6I3</td>\n      <td>A0A096NSB3</td>\n      <td>36</td>\n      <td>51</td>\n      <td>35</td>\n      <td>50</td>\n      <td>2</td>\n      <td>2</td>\n      <td>2</td>\n      <td>2</td>\n      <td>ASLTCTFSGGINVAGY</td>\n      <td>TARLPCTLSSDLSVGG</td>\n    </tr>\n    <tr>\n      <th>...</th>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n    </tr>\n    <tr>\n      <th>19007</th>\n      <td>19007</td>\n      <td>Q6UXU4</td>\n      <td>F7ILG0</td>\n      <td>4</td>\n      <td>37</td>\n      <td>0</td>\n      <td>30</td>\n      <td>2</td>\n      <td>2</td>\n      <td>1</td>\n      <td>1</td>\n      <td>ELAKAFSSQRTRLSAILSMLSLSFSTISLFSNYW</td>\n      <td>MKTSRRGRALLAVALNLLALLFATTAFLTTH</td>\n    </tr>\n    <tr>\n      <th>19008</th>\n      <td>19008</td>\n      <td>Q6UXU4</td>\n      <td>F7ILG0</td>\n      <td>44</td>\n      <td>77</td>\n      <td>37</td>\n      <td>83</td>\n      <td>2</td>\n      <td>2</td>\n      <td>1</td>\n      <td>1</td>\n      <td>VPKPLCEKDLAAKCFDMPVSLDGDVANTSTQEVV</td>\n      <td>RVPKPGCGQGGRANCPNSGANATANGTAAPAAAAAAATASGNGPPGG</td>\n    </tr>\n    <tr>\n      <th>19009</th>\n      <td>19009</td>\n      <td>Q6UXU4</td>\n      <td>F7ILG0</td>\n      <td>115</td>\n      <td>245</td>\n      <td>121</td>\n      <td>273</td>\n      <td>3</td>\n      <td>5</td>\n      <td>2</td>\n      <td>5</td>\n      <td>FIELTPPTERGEKGLLEFATLQGPCHPTLRFGGKQLMEKASLPYPP...</td>\n      <td>SFIDLAPASEKGVLWLSVVSEVLYILLLVVGFSLMCLELFHSSNVI...</td>\n    </tr>\n    <tr>\n      <th>19010</th>\n      <td>19010</td>\n      <td>Q6UXU4</td>\n      <td>F7ILG0</td>\n      <td>247</td>\n      <td>286</td>\n      <td>276</td>\n      <td>327</td>\n      <td>5</td>\n      <td>6</td>\n      <td>5</td>\n      <td>7</td>\n      <td>LELRLGLLHGLAFLHLLHGVGGHHLQYVHQDGAGVQVQA</td>\n      <td>RMEKRDGSEEDFHLDCRHERYPARHQPHMADSWPRSSAQEAPELNR...</td>\n    </tr>\n    <tr>\n      <th>19011</th>\n      <td>19011</td>\n      <td>Q6UXU6</td>\n      <td>H0XV48</td>\n      <td>80</td>\n      <td>95</td>\n      <td>79</td>\n      <td>96</td>\n      <td>4</td>\n      <td>4</td>\n      <td>5</td>\n      <td>5</td>\n      <td>RSCIQPEQRNLQESSE</td>\n      <td>CRNCREPEPDSPVDCRGP</td>\n    </tr>\n  </tbody>\n</table>\n<p>19012 rows × 13 columns</p>\n</div>",
      "text/plain": "       mismatch_ID    prot_hum   prot_prim  pos_start_prim  pos_stop_prim  \\\n0                0  A0A044PY82      H0WKS4               1             55   \n1                1  A0A044PY82      H0WKS4             108            119   \n2                2  A0A044PY82      G3QIT3              92            111   \n3                3  A0A075B6I0  A0A0D9RMQ9              74             97   \n4                4  A0A075B6I3  A0A096NSB3              36             51   \n...            ...         ...         ...             ...            ...   \n19007        19007      Q6UXU4      F7ILG0               4             37   \n19008        19008      Q6UXU4      F7ILG0              44             77   \n19009        19009      Q6UXU4      F7ILG0             115            245   \n19010        19010      Q6UXU4      F7ILG0             247            286   \n19011        19011      Q6UXU6      H0XV48              80             95   \n\n       pos_start_hum  pos_stop_hum  exon_start_prim  exon_stop_prim  \\\n0                  0            50                1               3   \n1                110           122                3               3   \n2                 91           119                2               3   \n3                 78           117                2               2   \n4                 35            50                2               2   \n...              ...           ...              ...             ...   \n19007              0            30                2               2   \n19008             37            83                2               2   \n19009            121           273                3               5   \n19010            276           327                5               6   \n19011             79            96                4               4   \n\n       exon_start_hum  exon_stop_hum  \\\n0                   1              2   \n1                   2              2   \n2                   2              2   \n3                   2              2   \n4                   2              2   \n...               ...            ...   \n19007               1              1   \n19008               1              1   \n19009               2              5   \n19010               5              7   \n19011               5              5   \n\n                                                seq_prim  \\\n0      AGNPAQPSPGRESLARSKAAPGAPSGLSCAQQGCRGRRPALPLGGR...   \n1                                           KISSGSSLETPQ   \n2                                   GPSDRSAKGGGPGAASPSSP   \n3                               CPSGVPESLLWLHPWEQSCPHHHG   \n4                                       ASLTCTFSGGINVAGY   \n...                                                  ...   \n19007                 ELAKAFSSQRTRLSAILSMLSLSFSTISLFSNYW   \n19008                 VPKPLCEKDLAAKCFDMPVSLDGDVANTSTQEVV   \n19009  FIELTPPTERGEKGLLEFATLQGPCHPTLRFGGKQLMEKASLPYPP...   \n19010            LELRLGLLHGLAFLHLLHGVGGHHLQYVHQDGAGVQVQA   \n19011                                   RSCIQPEQRNLQESSE   \n\n                                                 seq_hum  \n0      MAPQRAGPGLGKPSSKKSRALRGGPGVPRARQRSPRAAAELNTEVP...  \n1                                          PKTSSSRSGSEAP  \n2                          PGPSEAQRAEVQGQHPPPGPKTSSSRSGS  \n3               TRSSGVPDRFSGSILGNKAALTITGAQADDESDYYCVLYM  \n4                                       TARLPCTLSSDLSVGG  \n...                                                  ...  \n19007                    MKTSRRGRALLAVALNLLALLFATTAFLTTH  \n19008    RVPKPGCGQGGRANCPNSGANATANGTAAPAAAAAAATASGNGPPGG  \n19009  SFIDLAPASEKGVLWLSVVSEVLYILLLVVGFSLMCLELFHSSNVI...  \n19010  RMEKRDGSEEDFHLDCRHERYPARHQPHMADSWPRSSAQEAPELNR...  \n19011                                 CRNCREPEPDSPVDCRGP  \n\n[19012 rows x 13 columns]"
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mismatch"
   ]
  }
 ]
}